---
title: "gene_signatures"
author: "Kent Riemondy RBI"
date: "`r Sys.Date()`"
knit: (function(inputFile, encoding) {
  rmarkdown::render(inputFile, encoding = encoding, output_dir = "html") })
output:
  html_document:
    toc: true
    toc_float: true
    toc_collapsed: false
    theme: cosmo
    highlight: tango
    fig_caption: true
    code_folding: hide
    df_print: paged
---


```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r libs}
source(here::here("R/utils.R"))
library(ComplexHeatmap)
library(presto)
library(openxlsx)
library(gprofiler2)
library(ggrepel)
library(igraph)
library(GeneOverlap)
```


```{r}

so_fns <- file.path(obj_dir, c(
  file.path("shh", "shh.qs"),
  file.path("gp34", "gp3_only.qs"),
  file.path("gp34", "gp4_only.qs")))
                    
clusters <- c(
  "tumor_subpopulations_shh",
  "tumor_subpopulations_gp3",
  "tumor_subpopulations_gp4"
)

sos <- map(so_fns, 
            ~qread(.x, nthreads = 4, use_alt_rep = FALSE))

names(sos) <- c("SHH", "GP3", "GP4")

```

```{r}
ortholog_table <- "../dbases/mouse_orthologs.tsv"

if(!file.exists(ortholog_table)){
  library(biomaRt)
  ensembl <- useMart("ensembl", dataset = "hsapiens_gene_ensembl")
  
  orthologs <- getBM(mart = ensembl, 
      attributes = c("external_gene_name",
                     "mmusculus_homolog_associated_gene_name"))
  
  write_tsv(orthologs, ortholog_table)
}

orthologs <- read_tsv(ortholog_table)

orthologs <- filter(orthologs,
                            !is.na(mmusculus_homolog_associated_gene_name)) %>% 
  group_by(external_gene_name) %>% 
  mutate(n_orthos_h = n()) %>% 
  group_by(mmusculus_homolog_associated_gene_name) %>% 
  mutate(n_orthos_m = n()) %>% 
  filter(n_orthos_h == 1, n_orthos_m == 1) %>% 
  dplyr::select(-starts_with("n_orthos")) %>% 
  ungroup()

orthologs <- mutate(orthologs, 
                    ortho_id = str_c(external_gene_name, ":",
                                     mmusculus_homolog_associated_gene_name))
```

## Regenerate marker tables etc.

Using the new cell type labels

```{r}
all_mrkrs <- pmap(
  list(sos,
       names(sos),
       clusters
  ),
  function(so, id, cluster){
    
    mk_dir <- file.path(mkrs_dir, id)
    dir.create(mk_dir, showWarnings = FALSE)
    tbl_dir <- file.path(tbls_dir, id)
    dir.create(tbl_dir, showWarnings = FALSE)
    
    # markers
    full_mkrs <-  wilcoxauc(so, cluster) 
    
    mkrs <- filter(full_mkrs, logFC > 0, padj < 0.05, pct_in > 0.10) %>% 
      group_by(group) %>% 
      arrange(padj, desc(logFC), .by_group = TRUE)
    
    mkrs %>% 
      write_tsv(file.path(mk_dir, str_c("harmony_subpopulation_markers_", id, ".tsv")))
    
    mkrs %>% 
      ungroup() %>% 
      split(., .$group) %>% 
      write_markers_xlsx(., file.path(tbl_dir,
                                      str_c("harmony_subpopulation_markers_", id, ".xlsx")))
    
    # Go terms using top 500 markers 
    mkrs <- mkrs %>% 
      group_by(group) %>% 
      slice(1:500)  
    
    mkrs_split <- split(mkrs, mkrs$group) %>% 
      map(~pull(.x, feature))
    
    go_res <- gost(mkrs_split, 
                   ordered_query = TRUE,
                   sources = c("GO", 
                               "KEGG", 
                               "REAC", 
                               "TF"))
    
    go_res[["result"]] %>% 
      mutate(parents = unlist(map(parents, ~str_c(.x, collapse = ",")))) %>% 
      write_tsv(., file.path(tbl_dir, str_c("subpopulation_goterms_", id, ".tsv")))
    
    go_res[["result"]] %>% 
      mutate(parents = unlist(map(parents, ~str_c(.x, collapse = ","))),
             source = str_replace_all(source, "[[:punct:]]", " ")) %>% 
      select(-significant) %>% 
      split(., .$query) %>% 
      openxlsx::write.xlsx(., file.path(tbl_dir, 
                                        str_c("subpopulation_goterms_", id, ".xlsx")))
    
    ## Average Expression
    write_avg_expr(so,  
                   cluster,
                   file.path(tbl_dir, 
                             str_c("subpopulation_avgexpr_", id, ".csv")))
    ## Cell counts
    cell_count_mat <- get_cell_count_matrix(so, "UPN", cluster)
    cell_count_mat[is.na(cell_count_mat)] <- 0L
    cell_count_mat %>% 
      tibble::rownames_to_column("UPN") %>% 
      write_csv(file.path(tbl_dir, 
                          str_c("subpopulation_cell_counts_", id, ".csv")))
    
    ## marker matrix 
    full_marker_matrix <- full_mkrs %>%  top_marker_matrix(.) 
    
    topn_marker_matrix <- full_mkrs %>% top_marker_matrix(n = 200) 
    
    out <- list(avg_expression = read_csv(file.path(tbl_dir, 
                                                    str_c("subpopulation_avgexpr_", 
                                                          id,
                                                          ".csv"))),
                logFC_all = full_marker_matrix,
                top_200_logFC = topn_marker_matrix)
    
    openxlsx::write.xlsx(out, 
                         file.path(tbl_dir, 
                                   str_c("subpopulation_expression_logfc_summaries_",
                                         id,
                                         ".xlsx")))
    
    #full count matrix
    GetAssayData(so, "counts") %>% 
      as.matrix(.) %>% 
      as.data.frame() %>% 
      rownames_to_column("gene") %>% 
      data.table::fwrite(., file.path(tbl_dir, str_c(id, 
                                                     "_count_matrix.csv.gz")),
                         sep = ",",
                         compress = "gzip")
    
    # expression matrix filtered at least 25 cells express the gene
    expr_mat <- GetAssayData(so, "data")
    expr_mat <- expr_mat[Matrix::rowSums(expr_mat > 0) >= 25, ]
    
    expr_mat %>% 
      as.matrix(.) %>% 
      as.data.frame() %>% 
      rownames_to_column("gene") %>% 
      data.table::fwrite(., file.path(tbl_dir, str_c(id, "_expr_matrix.csv.gz")),
                         sep = ",",
                         compress = "gzip")
    
    ## TF markers
    Idents(so) <- cluster
    DefaultAssay(so) <- "TF"
    
    so_tf_markers <- wilcoxauc(so,
                               group_by = cluster, 
                               seurat_assay = "TF") %>% 
      filter(logFC > 0, 
             padj < 0.05) %>% 
      group_by(group) %>% 
      arrange(padj, desc(logFC))
    
    write_tsv(so_tf_markers, file.path(tbl_dir, 
                                       str_c("subpopulation_tf_activities_", id, ".tsv")))
    
    write_markers_xlsx(split(so_tf_markers, so_tf_markers$group),
                       path = file.path(tbl_dir, 
                                        str_c("subpopulation_tf_activities_", id, ".xlsx")),
                       description_string = "Transcription factor regulons derived from PyScenic that are significantly enriched in each cluster")
    
    mkrs
  })


```

```{r}

rmd_dir <- "cell_type_signature_comparisons"

fig_dir <- file.path(fig_dir, rmd_dir)
mkrs_dir <- file.path(mkrs_dir, rmd_dir)
tbls_dir <- file.path(tbls_dir, rmd_dir)
go_dir <- file.path(tbls_dir, "go")
xcel_markers_dir <- file.path(tbls_dir, "markers")

walk(c(fig_dir, mkrs_dir, tbls_dir, go_dir, xcel_markers_dir),
     dir.create, 
     showWarnings = FALSE)
```


```{r}
sos <- pmap(
  list(sos,
    names(sos),
    clusters),
            function(so, id, cluster){
              # drop the excluded populations
              to_keep <- rownames(so@meta.data)[!str_detect(so[[cluster]][,1], "-X[0-9]$")]
              so <- subset(so, cells = to_keep)
              so$tumor_cell_type <- so[[cluster]][,1]
              so
            })

avg_expr <- pmap(list(
  sos,
  names(sos),
  clusters),
           function(x, id, cluster){
            Idents(x) <- "tumor_cell_type"
             mat <- AverageExpression(x)$RNA
             x <- FindVariableFeatures(x, nfeatures = 2000)
             var_genes <- VariableFeatures(x)
             list(expr = mat,
                  var_genes = var_genes)
           })


var_genes <- map(avg_expr, ~.x$var_genes) %>% 
  unlist() %>% 
  unique()

avg_expr_mat <- map(avg_expr, ~.x$expr) %>% 
  unname() %>% 
  do.call(cbind, .)

avg_expr_mat <- avg_expr_mat[var_genes, ]


hmap <- Heatmap(cor(avg_expr_mat, method = "pearson"),
        col = viridis::viridis(256),
        name = "Correlation",
        row_names_gp = gpar(fontsize = 7),
        column_names_gp = gpar(fontsize = 7))

pdf(file.path(fig_dir, "tumor_subpopulation_correlation_matrix.pdf"),
    width = 9,
    height = 7)
draw(hmap)
dev.off()

hmap
```

## Correlate all subpopulations based on modules scores
```{r}
so <- merge(sos[[1]], sos[2:3])

dir_ids <- c("GP3", "shh", "GP4")
all_markers <- map(file.path(
  "markers",
  dir_ids,
  paste0("harmony_subpopulation_markers_", toupper(dir_ids), ".tsv")),
  ~read_tsv(.x) %>% 
    group_by(group) %>% 
    arrange(padj, desc(logFC), .by_group = T) %>% 
    ungroup()
)
  
var_marker_lst <- all_markers %>%
  map(~group_by(.x, group) %>% 
        slice(1:200)) %>% 
  bind_rows() %>% 
  split(., .$group) %>% 
  map(~pull(.x, feature))

var_marker_lst <- var_marker_lst[!str_detect(names(var_marker_lst),  "-X[0-9]$")]


for(i in seq_along(var_marker_lst)){
  mrks <- var_marker_lst[[i]]
  id <- names(var_marker_lst)[i]
  so <- AddModuleScore(so, 
                       list(x = mrks), 
                       assay = "RNA", 
                       name = id, 
                       seed = 42)
}

# rename cols
col_idxs <- match(str_c(make.names(names(var_marker_lst)), "1"), 
                    colnames(so@meta.data))
colnames(so@meta.data)[col_idxs] <- names(var_marker_lst)

all_tumor_cell_expression_scores <- get_metadata(so, embedding = NULL) %>% 
  .[, c("cell", names(var_marker_lst))] %>% 
  column_to_rownames("cell") %>% 
  as.matrix()

hmap <- Heatmap(cor(all_tumor_cell_expression_scores, method = "spearman"),
         col = viridis::viridis(256),
        name = "Correlation",
        row_names_gp = gpar(fontsize = 12),
        column_names_gp = gpar(fontsize = 12))

pdf(file.path(fig_dir, "tumor_subpopulation_scores_all_tumor_cells_cor_matrix.pdf"),
    width = 9,
    height = 7)
  draw(hmap)
dev.off()

hmap
```


## Make graphs based on correlation 

```{r}
so_gp3 <- qread(file.path("objects", "gp34", "gp3_only.qs"))
so_gp4 <- qread(file.path("objects", "gp34", "gp4_only.qs"))
so_shh <- qread(file.path("objects", "shh", "shh.qs")) 
  

gp4_to_keep <- colnames(so_gp4)[!str_detect(so_gp4$tumor_subpopulations_gp4, "X") ]
so_gp4 <- subset(so_gp4, cells = gp4_to_keep)

shh_to_keep <- colnames(so_shh)[so_shh$tumor_subpopulations_shh != "SHH-X1"]
so_shh <- subset(so_shh, cells = shh_to_keep)

sos <- list(
  gp3 = so_gp3,
  gp4 = so_gp4,
  shh = so_shh
)

rm(so_gp3, so_gp4, so_shh)
```

```{r}
avg_objs <- map2(
  sos,
  c("tumor_subpopulations_gp3",
    "tumor_subpopulations_gp4",
    "tumor_subpopulations_shh"),
  function(x, cluster_col) {
    Idents(x) <- cluster_col  
    avg_expr <- AverageExpression(x, return.seurat = FALSE)$RNA
    var_features <- VariableFeatures(FindVariableFeatures(x, nfeatures = 2000))
    list(expr = avg_expr, var_genes = var_features)
  })
```

```{r}
var_genes <- map(avg_objs, ~.x$var_genes) %>% 
  unlist() %>% 
  unique()

var_genes <- Reduce(intersect, map(avg_objs, ~.x$var_genes))

avg_expr_mat <- map(avg_objs, ~.x$expr) %>% 
  unname() %>% 
  do.call(cbind, .)

avg_expr_mat <- avg_expr_mat[var_genes, ]


hmap <- Heatmap(cor(avg_expr_mat, method = "pearson"),
        col = viridis::viridis(256),
        name = "Correlation",
        row_names_gp = gpar(fontsize = 7),
        column_names_gp = gpar(fontsize = 7))

pdf(file.path(fig_dir, "tumor_subpopulation_correlation_matrix.pdf"),
    width = 9,
    height = 7)
draw(hmap)
dev.off()
```


```{r}
pairwise_comps <- list(
  c("gp3", "gp4"),
  c("shh", "gp3"),
  c("shh", "gp4")
)

get_cor_values <- function(x, y, 
                           gene_combine_function = intersect,
                           cor_method = "spearman"){ 
  shared_var_genes <- gene_combine_function(x$var_genes, y$var_genes)
  
  x_mat <- x$expr[shared_var_genes, ] 
  
  y_mat <-  y$expr[shared_var_genes, ] 
  
  res <- list()
  res$r <- matrix(nrow = ncol(y_mat),
                  ncol = ncol(x_mat))
  res$pvals <- res$r
  
  for(m in seq_along(colnames(y_mat))){
    for(h in seq_along(colnames(x_mat))){
      s <- cor.test(y_mat[, m], x_mat[, h], method = cor_method)
      res$r[m, h] <- s$estimate
      res$pvals[m,h] <- s$p.value
    }
  }
  
  res$padj <- matrix(p.adjust(res$pvals),
                     nrow = nrow(res$pvals), 
                     ncol = ncol(res$pvals),
                     byrow = FALSE)
  res <- map(res, ~{
    rownames(.x) <- colnames(y_mat)
    colnames(.x) <- colnames(x_mat)
    .x
  })
  
  res
}
```


### GP3 vs. GP4

```{r}
res <- get_cor_values(avg_objs[["gp3"]], avg_objs[["gp4"]])
seed_val <- 20200613

# weighted graph based on correlation
g <- graph.incidence(res$r, weighted = TRUE)

# prune edges with correlation < 0.8 similar to clustifyr (or non-sig)
g <- delete_edges(g, which(E(g)$weight <= max(res$r) * 0.8 | as.vector(t(res$padj)) >= 0.05))

# add some meta clusters
set.seed(seed_val)
wc <- cluster_walktrap(g)

V(g)$color <- V(g)$type
V(g)$shape <- ifelse(V(g)$type, "circle", "rectangle")

# force directed layout
set.seed(seed_val)
layout <- layout.fruchterman.reingold(g)

new_cols <- ifelse(startsWith(V(g)$name, "GP3"), "white", "black")

min_cor <- min(E(g)$weight)
max_cor <- max(E(g)$weight)
plot(wc, g,
     col = new_cols,
     mark.col = alpha(palette_OkabeIto[1:length(groups(wc))], 0.75),
     mark.border = "black",
     layout = layout,
    edge.width= scales::rescale(E(g)$weight, c(1,5)), 
     vertex.label = str_replace(V(g)$name, "-", "\n"),
     vertex.label.family = "sans",
    vertex.label.color = ifelse(startsWith(V(g)$name, "GP3"), "black", "white"))

legend("bottomleft",
       legend = c(paste("r =", formatC(min_cor, digits = 2)),
                  paste("r =", formatC(max_cor, digits = 2))),
       lwd = c(min(scales::rescale(E(g)$weight, c(1,5))),
               max(scales::rescale(E(g)$weight, c(1,5)))))
       

pdf(file.path(fig_dir,
              "gp3_to_gp4_correlation_graph.pdf"),
    height = 9, width = 9)
plot(wc, g,
     col = new_cols,
     mark.col = alpha(palette_OkabeIto[1:length(groups(wc))], 0.75),
     mark.border = "black",
     layout = layout,
    edge.width= scales::rescale(E(g)$weight, c(1,5)), 
     vertex.label = str_replace(V(g)$name, "-", "\n"),
     vertex.label.family = "sans",
    vertex.label.color = ifelse(startsWith(V(g)$name, "GP3"), "black", "white"))

legend("bottomleft",
       legend = c(paste("r =", formatC(min_cor, digits = 2)),
                  paste("r =", formatC(max_cor, digits = 2))),
       lwd = c(min(scales::rescale(E(g)$weight, c(1,5))),
               max(scales::rescale(E(g)$weight, c(1,5)))))
       
dev.off()
```


### GP3 vs. SHH

```{r}
res <- get_cor_values(avg_objs[["gp3"]], avg_objs[["shh"]])
seed_val <- 20200613

# weighted graph based on correlation
g <- graph.incidence(res$r, weighted = TRUE)

# prune edges with correlation < 0.8 similar to clustifyr (or non-sig)
g <- delete_edges(g, which(E(g)$weight <= max(res$r) * 0.8 | as.vector(t(res$padj)) >= 0.05))

# add some meta clusters
set.seed(seed_val)
wc <- cluster_walktrap(g)

V(g)$color <- V(g)$type
V(g)$shape <- ifelse(V(g)$type, "circle", "rectangle")

# force directed layout
set.seed(seed_val)
layout <- layout.fruchterman.reingold(g)

new_cols <- ifelse(startsWith(V(g)$name, "GP3"), "white", "black")

min_cor <- min(E(g)$weight)
max_cor <- max(E(g)$weight)
plot(wc, g,
     col = new_cols,
     mark.col = alpha(palette_OkabeIto[1:length(groups(wc))], 0.75),
     mark.border = "black",
     layout = layout,
    edge.width= scales::rescale(E(g)$weight, c(1,5)), 
     vertex.label = str_replace(V(g)$name, "-", "\n"),
     vertex.label.family = "sans",
    vertex.label.color = ifelse(startsWith(V(g)$name, "GP3"), "black", "white"))

legend("bottomleft",
       legend = c(paste("r =", formatC(min_cor, digits = 2)),
                  paste("r =", formatC(max_cor, digits = 2))),
       lwd = c(min(scales::rescale(E(g)$weight, c(1,5))),
               max(scales::rescale(E(g)$weight, c(1,5)))))
       

pdf(file.path(fig_dir,
              "gp3_to_shh_correlation_graph.pdf"),
    height = 9, width = 9)
plot(wc, g,
     col = new_cols,
     mark.col = alpha(palette_OkabeIto[1:length(groups(wc))], 0.75),
     mark.border = "black",
     layout = layout,
    edge.width= scales::rescale(E(g)$weight, c(1,5)), 
     vertex.label = str_replace(V(g)$name, "-", "\n"),
     vertex.label.family = "sans",
    vertex.label.color = ifelse(startsWith(V(g)$name, "GP3"), "black", "white"))

legend("bottomleft",
       legend = c(paste("r =", formatC(min_cor, digits = 2)),
                  paste("r =", formatC(max_cor, digits = 2))),
       lwd = c(min(scales::rescale(E(g)$weight, c(1,5))),
               max(scales::rescale(E(g)$weight, c(1,5)))))
       
dev.off()
```



```{r}
library("GeneOverlap")

mouse_markers <- read_tsv(file.path("markers", "mouse", "gp34", "myc_gfi1_malignant_cluster_markers.tsv")) %>% 
filter(padj < 0.05) %>% 
  group_by(group) %>% 
  arrange(padj, desc(logFC), .by_group = TRUE) %>% 
  inner_join(orthologs, by = c("feature" = "mmusculus_homolog_associated_gene_name")) %>% 
    slice(1:200) %>% 
  split(., .$group) %>% 
  map(~pull(.x, external_gene_name))

human_markers <- read_tsv(file.path("markers", "GP3",  "harmony_subpopulation_markers_GP3.tsv")) %>% 
  filter(padj < 0.05) %>% 
  group_by(group) %>% 
  arrange(padj, desc(logFC), .by_group = TRUE) %>% 
  semi_join(orthologs, by = c("feature" = "external_gene_name")) %>% 
  slice(1:200) %>% 
  split(., .$group) %>% 
  map(~pull(.x, feature))


gom_human_mouse <- newGOM(mouse_markers, 
                       human_markers,
                       genome.size = nrow(nrow(sos$gp3)))

# extract overlaps
pvals  <- getMatrix(gom_human_mouse, c("pval")) 
pvals <- apply(pvals, 2, function(x) p.adjust(x, n = length(pvals)))

jaccard <- getMatrix(gom_human_mouse, "Jaccard")
odds <- getMatrix(gom_human_mouse, c("odds.ratio"))

library(ComplexHeatmap)
# clip really low pvals to avoid take log10 of 0)
clip_val <- 1e-300
  
pvals[pvals < clip_val] <- clip_val
pvals <- -log10(pvals)

Heatmap(pvals,
        name = "-log10(pvalue)", 
        col = viridis::viridis(256),
        row_title = "Mouse model markers",
        column_title = "Human Group 3 and 4 markers")

h <- Heatmap(jaccard,
        name = "Jaccard index", 
        col = viridis::viridis(256),
       row_title = "Myc + GFI1 model clusters",
        column_title = "Human Group 3 and 4 clusters",
        cell_fun = function(j, i, x, y, width, height, fill) {
          bin <- findInterval(pvals[i, j], c(-log10(0.01), 20, 200))
          txt <- c("*", "**", "***")[bin]
          grid.text(txt, x, y, gp = gpar(fontsize = 10, col = "white"))
})


lgd_list <-  list(
    Legend(labels = c("*** < 1e-200",
                      "** < 1e-20",
                      "* < 1e-2"), 
           title = "Adj. p.val")
)
draw(h, heatmap_legend_list = lgd_list)


pdf(file.path(fig_dir, "Marker_overlap_gfi1_malignant_gp34.pdf"),
    width = 7,
    height = 7)
  draw(h, heatmap_legend_list = lgd_list)
dev.off()
```


### GP4 vs. SHH

```{r}
res <- get_cor_values(avg_objs[["gp4"]], avg_objs[["shh"]])
seed_val <- 20200613


# weighted graph based on correlation
g <- graph.incidence(res$r, weighted = TRUE)

# prune edges with correlation < 0.8 similar to clustifyr (or non-sig)
g <- delete_edges(g, which(E(g)$weight <= max(res$r) * 0.8 | as.vector(t(res$padj)) >= 0.05))

# add some meta clusters
set.seed(seed_val)
wc <- cluster_walktrap(g)

V(g)$color <- V(g)$type
V(g)$shape <- ifelse(V(g)$type, "circle", "rectangle")

# force directed layout
set.seed(seed_val)
layout <- layout.fruchterman.reingold(g)

new_cols <- ifelse(startsWith(V(g)$name, "GP4"), "white", "black")

min_cor <- min(E(g)$weight)
max_cor <- max(E(g)$weight)
plot(wc, g,
     col = new_cols,
     mark.col = alpha(palette_OkabeIto[1:length(groups(wc))], 0.75),
     mark.border = "black",
     layout = layout,
    edge.width= scales::rescale(E(g)$weight, c(1,5)), 
     vertex.label = str_replace(V(g)$name, "-", "\n"),
     vertex.label.family = "sans",
    vertex.label.color = ifelse(startsWith(V(g)$name, "GP4"), "black", "white"))

legend("bottomleft",
       legend = c(paste("r =", formatC(min_cor, digits = 2)),
                  paste("r =", formatC(max_cor, digits = 2))),
       lwd = c(min(scales::rescale(E(g)$weight, c(1,5))),
               max(scales::rescale(E(g)$weight, c(1,5)))))
       

pdf(file.path(fig_dir,
              "gp4_to_shh_correlation_graph.pdf"),
    height = 9, width = 9)
plot(wc, g,
     col = new_cols,
     mark.col = alpha(palette_OkabeIto[1:length(groups(wc))], 0.75),
     mark.border = "black",
     layout = layout,
    edge.width= scales::rescale(E(g)$weight, c(1,5)), 
     vertex.label = str_replace(V(g)$name, "-", "\n"),
     vertex.label.family = "sans",
    vertex.label.color = ifelse(startsWith(V(g)$name, "GP4"), "black", "white"))

legend("bottomleft",
       legend = c(paste("r =", formatC(min_cor, digits = 2)),
                  paste("r =", formatC(max_cor, digits = 2))),
       lwd = c(min(scales::rescale(E(g)$weight, c(1,5))),
               max(scales::rescale(E(g)$weight, c(1,5)))))
       
dev.off()
```



### Mouse models -> Human tumors
#### Combine mouse models

```{r}
so_sub_p53 <- qread(file.path(obj_dir,"mouse", "gp34", "gp3_myc_dnp53_sobj.qs")) %>% 
  subset(., subset = cell_types %in% c("malignant"))

so_sub_gfi <- qread(file.path(obj_dir,"mouse", "gp34", "gp3_myc_gfi1_sobj.qs")) %>% 
                      subset(., 
                 subset = cell_types %in% c("malignant", "Neuroendocrine"))

so_shh_sub <- qread(file.path(obj_dir,"mouse", "shh", "shh_sobj.qs")) %>% 
  subset(., subset = cell_types == "malignant")

m_sos <- list(
  dnp53_myc = so_sub_p53,
  gfi1_myc = so_sub_gfi,
  shh = so_shh_sub
)


m_sos <- imap(m_sos, 
             ~{.x$clusters <- paste0(.y, "_", .x$clusters)
               .x
               })
m_so <- Reduce(function(x, y) merge(x, y), m_sos)
rm(m_sos, so_sub_p53, so_sub_gfi, so_shh_sub)
```

#### Combine human tumors

```{r}
so_gp3 <- qread(file.path("objects", "gp34", "gp3_only.qs"))
so_gp4 <- qread(file.path("objects", "gp34", "gp4_only.qs"))
so_shh <- qread(file.path("objects", "shh", "shh.qs")) 
 
so_gp3$subgroup <- "GP3"
gp4_to_keep <- colnames(so_gp4)[!str_detect(so_gp4$tumor_subpopulations_gp4, "X") ]
so_gp4 <- subset(so_gp4, cells = gp4_to_keep)

shh_to_keep <- colnames(so_shh)[so_shh$tumor_subpopulations_shh != "SHH-X1"]
so_shh <- subset(so_shh, cells = shh_to_keep)

h_sos <- list(
  gp3 = so_gp3,
  gp4 = so_gp4,
  shh = so_shh
)

h_sos$gp3$subpopulation <- h_sos$gp3$tumor_subpopulations_gp3
h_sos$gp4$subpopulation <- h_sos$gp4$tumor_subpopulations_gp4
h_sos$shh$subpopulation <- h_sos$shh$tumor_subpopulations_shh

h_so <- Reduce(function(x, y) merge(x, y), h_sos)
rm(h_sos, so_gp3, so_gp4, so_shh)
```



```{r}
mats <- set_shared_orthologs(h_so, m_so, orthologs)
names(mats) <- c("human", "mouse")
  
sos <- pmap(list(
  mats, 
  list(h_so, m_so),
  c("subpopulation",
    "clusters")),
  function(mat, so, cluster_col) {
  tmp <- CreateSeuratObject(mat, meta.data = so@meta.data)
  Idents(tmp) <- cluster_col  
  avg_expr <- AverageExpression(tmp, return.seurat = FALSE)$RNA
  
  var_features <- VariableFeatures(FindVariableFeatures(tmp, nfeatures = 4000))
  
  list(avg_expr, var_features, tmp)
})
```

#### Make integrated UMAP

```{r}
pretty_mouse_model_names <- c(
  "SHH" = "SHH model (SmoM2/Math1-Cre)",
  "MYC + DNP53" = "GP3 model (Myc-T58A/dn-Trp53)",
  "MYC + GFI1" = "GP3 model (Myc-T58A/Gfi1)"
)

sos$human[[3]]$species <- "human"
sos$mouse[[3]]$species <- "mouse"
sos$mouse[[3]]$model <- ifelse(startsWith(sos$mouse[[3]]$UPN, "MED"),
                               "SHH",
                               sos$mouse[[3]]$expt_ids)

sos$mouse[[3]]$model <- pretty_mouse_model_names[sos$mouse[[3]]$model]

sos$mouse[[3]]$subgroup <- sos$mouse[[3]]$model
sos$mouse[[3]]$subpopulation <- sos$mouse[[3]]$clusters

so <-  merge(sos$human[[3]], sos$mouse[[3]])

so <- FindVariableFeatures(so, nfeatures = 4000) %>% 
  ScaleData() %>% 
  RunPCA(npcs = 50) 

so <- RunUMAP(so, dims = 1:50)

library(harmony)
so <- RunHarmony(so, 
                 "species",
                 theta = 1.0, 
                 reduction.save = "harmony_1") %>% 
  RunUMAP(dims = 1:50,
          reduction = "harmony_1", 
          reduction.name = "Harmony_UMAP_1")

so <- RunHarmony(so, 
                 "species",
                 theta = 2.0, 
                 reduction.save = "harmony_2") %>% 
  RunUMAP(dims = 1:50, 
          reduction = "harmony_2", 
          reduction.name = "Harmony_UMAP_2")

so <- RunHarmony(so, 
                 "species",
                 theta = 3.0, 
                 reduction.save = "harmony_3") %>% 
  RunUMAP(dims = 1:50, 
          reduction = "harmony_3", 
          reduction.name = "Harmony_UMAP_3")


so$subgroup <- factor(so$subgroup, levels = c("GP3",
    "GP4",
    "SHH",
    "GP3 model (Myc-T58A/dn-Trp53)",
    "GP3 model (Myc-T58A/Gfi1)",
    "SHH model (SmoM2/Math1-Cre)"
))

small_labels <- c("GP3" = "GP3",
    "GP4" = "GP4",
    "SHH" = "SHH",
    "GP3 model (Myc-T58A/dn-Trp53)"  = "GP3 model MP",
    "GP3 model (Myc-T58A/Gfi1)" = "GP3 model MG",
    "SHH model (SmoM2/Math1-Cre)" = "SHH model MS"
)

so$subgroup_small <- small_labels[so$subgroup]
so$subgroup_small  <- factor(so$subgroup_small, levels = small_labels)
umap_types <- c("umap", "Harmony_UMAP_1", "Harmony_UMAP_2", "Harmony_UMAP_3")
umaps <- map(umap_types,
             ~plot_feature(so, 
                           "subgroup_small", 
                           .cols = palette_OkabeIto[-4],
                           sorted = "random",
                           legend_title = "", embedding = .x) +
               labs(x = "UMAP 1",
                    y = "UMAP 2")) 
```

```{r}
Idents(so) <- "subgroup_small"

subgroups <- unique(so$subgroup_small)
reductions <- c("pca", "harmony_1", "harmony_2", "harmony_3")
avg_h <- map(reductions, ~{
  h <- Embeddings(so, .x)
  
  avg_h <- lapply(subgroups, 
                  function(x){
                    ids <- names(so$subgroup_small)[so$subgroup_small == x]
                    h_tmp <- h[ids, ]
                    m <- colMeans(h_tmp)
                    df <- data.frame(row.names = names(m),
                                     tmp = m)
                    colnames(df) <- x
                    df
                  }) %>% 
    do.call(cbind, .) 
  avg_h
})

library(dendextend)

dends <- map(avg_h, ~{
  dend <- as.dendrogram(hclust(dist(t(.x)), method = "ward.D2"))
  labels(dend) <- str_replace(labels(dend), "model ", "model\n")
  labels_colors(dend) <- palette_OkabeIto[-4][order.dendrogram(dend)]
  dend <- set(dend, "labels_cex", 0.8)
  dend <- set(dend, "branches_lwd", 0.8)
  p <- dend %>% 
    hang.dendrogram %>% 
    ggplot(., yaxt = 'n', offset_labels = -1) +
    ylim(-30, max(get_branches_heights(dend))) 
})

plts <- map2(umaps, dends, ~list(.x, .y)) %>% flatten()
plt <- plot_grid(plotlist = plts,
                 nrow = 4, 
                 ncol = 2, 
                 rel_widths = c(1.5, 1), 
                 labels = c("", "   PCA          ",
                            "", "Harmony (theta = 1)",
                            "", "Harmony (theta = 2)", 
                            "", "Harmony (theta = 3)"),
                 hjust = 1)
save_plot(file.path(fig_dir,
                    "cell_type_signature_comparisons",
                    "mouse-human-harmonized-comparison-v2.pdf"),
          plt,
          base_asp = 1.3,
          nrow = 4, 
          ncol = 2)

# just plot theta 2 

umaps[[3]]

plt <- plot_grid(umaps[[3]], dends[[3]],
                 nrow = 1, 
                 ncol = 2, 
                 rel_widths = c(1.5, 1), 
                 labels = c("", "Harmony (theta = 2)"),
                 hjust = 1)
save_plot(file.path(fig_dir,
                    "cell_type_signature_comparisons",
                    "mouse-human-harmonized-comparison-theta2.pdf"),
          plt,
          base_asp = 1.3,
          nrow = 1, 
          ncol = 2)
```

#### Examine correlation between signatures

```{r}
shared_var_genes <- union(sos$human[[2]], sos$mouse[[2]])

h_var_mat <- sos$human[[1]][shared_var_genes, ] %>% 
  t() %>% 
  scale(scale = FALSE, center = TRUE) %>% 
  t()

m_var_mat <-  sos$mouse[[1]][shared_var_genes, ] %>% 
     t() %>% 
  scale(scale = FALSE, center = TRUE) %>% 
  t()

res <- list()
res$r <- matrix(nrow = ncol(m_var_mat),
                ncol = ncol(h_var_mat))
res$pvals <- res$r

for(m in seq_along(colnames(m_var_mat))){
  for(h in seq_along(colnames(h_var_mat))){
    s <- cor.test(m_var_mat[, m], h_var_mat[, h], method = "pearson")
    res$r[m, h] <- s$estimate
    res$pvals[m,h] <- s$p.value
  }
}

res$padj <- matrix(p.adjust(res$pvals),
                   nrow = nrow(res$pvals), 
                   ncol = ncol(res$pvals),
                   byrow = FALSE)
res <- map(res, ~{
  rownames(.x) <- colnames(m_var_mat)
  colnames(.x) <- colnames(h_var_mat)
  .x
})

library(igraph)
g <- graph.incidence(res$r, weighted = TRUE)
g <- delete_edges(g, which(E(g)$weight <= 0.2 | as.vector(t(res$padj)) >= 0.05))
wc <- cluster_walktrap(g)

V(g)$color <- V(g)$type
V(g)$shape <- ifelse(V(g)$type, "circle", "rectangle")
#plot(g, edge.width=E(g)$weight, layout=layout.fruchterman.reingold)

layout <-layout.fruchterman.reingold(g)

new_cols <- ifelse(str_detect(V(g)$name, "^[A-Z]"), 
                   "white",
                   ifelse(startsWith(V(g)$name, "shh"),
                          "black",
                          ifelse(startsWith(V(g)$name, "gfi1"),
                          "red",
                          "grey")))
                   
min_cor <- min(E(g)$weight)
max_cor <- max(E(g)$weight)
plot(wc, g,
     col = new_cols,
     mark.col = alpha(palette_OkabeIto[1:length(groups(wc))], 0.75),
     mark.border = "black",
     layout = layout,
    edge.width= scales::rescale(E(g)$weight, c(1,5)))
    #  vertex.label = ifelse(startsWith(V(g)$name, "GP3"), 
    #                        str_remove(V(g)$name, "GP3-"), 
    #                        str_split(V(g)$name, "-", simplify = T) %>% .[, 2] %>% str_c("M", .)),
    #  vertex.label.family = "sans",
    # vertex.label.color = ifelse(startsWith(V(g)$name, "GP3"), "black", "white"))

legend("bottomleft",
       legend = c(paste("r =", formatC(min_cor, digits = 2)),
                  paste("r =", formatC(max_cor, digits = 2))),
       lwd = c(min(scales::rescale(E(g)$weight, c(1,5))),
               max(scales::rescale(E(g)$weight, c(1,5)))))
       
legend("bottomright",
       legend = c("Human GP3 subpopulations",
                  "Mouse DNp53 GP3 model",
                  "Mouse Gfi1 GP3 model"),
       pch = c(21, 22, 22),
       pt.bg = c("white", "black", "grey"))

pdf(file.path(fig_dir, "gfi1_and_dnp53_mouse_to_human_correlation_graph.pdf"))
min_cor <- min(E(g)$weight)
max_cor <- max(E(g)$weight)
plot(wc, g,
     col = new_cols,
     mark.col = alpha(palette_OkabeIto[1:length(groups(wc))], 0.75),
     mark.border = "black",
     layout = layout,
    edge.width= scales::rescale(E(g)$weight, c(1,5)), 
     vertex.label = ifelse(startsWith(V(g)$name, "GP3"), 
                           str_remove(V(g)$name, "GP3-"), 
                           str_split(V(g)$name, "-", simplify = T) %>% .[, 2] %>% str_c("M", .)),
     vertex.label.family = "sans",
    vertex.label.color = ifelse(startsWith(V(g)$name, "GP3"), "black", "white"))

legend("bottomleft",
       legend = c(paste("r =", formatC(min_cor, digits = 2)),
                  paste("r =", formatC(max_cor, digits = 2))),
       lwd = c(min(scales::rescale(E(g)$weight, c(1,5))),
               max(scales::rescale(E(g)$weight, c(1,5)))),
       cex = 0.60)

legend("bottomright",
       legend = c("Human GP3 subpopulations",
                  "Mouse DNp53 GP3 model",
                  "Mouse Gfi1 GP3 model"),
       pch = c(21, 22, 22),
       pt.bg = c("white", "black", "grey"),
       cex = 0.60)
dev.off()
```


## Marker overlap 

### GP3 vs GP4 

```{r}
gp3_markers <- read_tsv(file.path("markers", "GP3", "harmony_subpopulation_markers_GP3.tsv")) %>% 
  filter(padj < 0.05,
         !str_detect(group, "X")) %>% 
  group_by(group) %>% 
  arrange(padj, desc(logFC), .by_group = TRUE) %>% 
  slice(1:200) %>% 
  split(., .$group) %>% 
  map(~pull(.x, feature))

gp4_markers <- read_tsv(file.path("markers", "GP4",  "harmony_subpopulation_markers_GP4.tsv")) %>% 
  filter(padj < 0.05,
         !str_detect(group, "X")) %>% 
  group_by(group) %>% 
  arrange(padj, desc(logFC), .by_group = TRUE) %>%
  slice(1:200) %>% 
  split(., .$group) %>% 
  map(~pull(.x, feature))


gom <- newGOM(gp3_markers, 
              gp4_markers,
              genome.size = nrow(sos$gp3))

# extract overlaps
pvals  <- getMatrix(gom, c("pval")) 
pvals <- apply(pvals, 2, function(x) p.adjust(x, n = length(pvals)))

jaccard <- getMatrix(gom, "Jaccard")
odds <- getMatrix(gom, c("odds.ratio"))


# clip really low pvals to avoid take log10 of 0)
clip_val <- 1e-300
  
pvals[pvals < clip_val] <- clip_val
pvals <- -log10(pvals)

h <- Heatmap(jaccard,
        name = "Jaccard index", 
        col = viridis::viridis(256),
       row_title = "Group 3 subpopulations",
        column_title = "Group 4 subpopulations",
        cell_fun = function(j, i, x, y, width, height, fill) {
          bin <- findInterval(pvals[i, j], c(-log10(0.01), 20, 200))
          txt <- c("*", "**", "***")[bin]
          grid.text(txt, x, y, gp = gpar(fontsize = 10, col = "white"))
})


lgd_list <-  list(
    Legend(labels = c("*** < 1e-200",
                      "** < 1e-20",
                      "* < 1e-2"), 
           title = "Adj. p.val")
)
draw(h, heatmap_legend_list = lgd_list)


pdf(file.path(fig_dir, "Marker_overlap_gp3_gp4.pdf"),
    width = 7,
    height = 7)
  draw(h, heatmap_legend_list = lgd_list)
dev.off()
```

### GP3 vs SHH 

```{r}
shh_markers <- read_tsv(file.path("markers", "SHH",  "harmony_subpopulation_markers_SHH.tsv")) %>% 
  filter(padj < 0.05,
         !str_detect(group, "X")) %>% 
  group_by(group) %>% 
  arrange(padj, desc(logFC), .by_group = TRUE) %>%
  slice(1:200) %>% 
  split(., .$group) %>% 
  map(~pull(.x, feature))


gom <- newGOM(gp3_markers, 
              shh_markers,
              genome.size = nrow(sos$gp3))

# extract overlaps
pvals  <- getMatrix(gom, c("pval")) 
pvals <- apply(pvals, 2, function(x) p.adjust(x, n = length(pvals)))

jaccard <- getMatrix(gom, "Jaccard")
odds <- getMatrix(gom, c("odds.ratio"))


# clip really low pvals to avoid take log10 of 0)
clip_val <- 1e-300
  
pvals[pvals < clip_val] <- clip_val
pvals <- -log10(pvals)

h <- Heatmap(jaccard,
        name = "Jaccard index", 
        col = viridis::viridis(256),
       row_title = "Group 3 subpopulations",
        column_title = "SHH subpopulations",
        cell_fun = function(j, i, x, y, width, height, fill) {
          bin <- findInterval(pvals[i, j], c(-log10(0.01), 20, 200))
          txt <- c("*", "**", "***")[bin]
          grid.text(txt, x, y, gp = gpar(fontsize = 10, col = "white"))
})


lgd_list <-  list(
    Legend(labels = c("*** < 1e-200",
                      "** < 1e-20",
                      "* < 1e-2"), 
           title = "Adj. p.val")
)
draw(h, heatmap_legend_list = lgd_list)


pdf(file.path(fig_dir, "Marker_overlap_gp3_shh.pdf"),
    width = 7,
    height = 7)
  draw(h, heatmap_legend_list = lgd_list)
dev.off()
```

### GP3 vs GP4 

```{r}
gom <- newGOM(gp4_markers, 
              shh_markers,
              genome.size = nrow(sos$gp4))

# extract overlaps
pvals  <- getMatrix(gom, c("pval")) 
pvals <- apply(pvals, 2, function(x) p.adjust(x, n = length(pvals)))

jaccard <- getMatrix(gom, "Jaccard")
odds <- getMatrix(gom, c("odds.ratio"))


# clip really low pvals to avoid take log10 of 0)
clip_val <- 1e-300
  
pvals[pvals < clip_val] <- clip_val
pvals <- -log10(pvals)

h <- Heatmap(jaccard,
        name = "Jaccard index", 
        col = viridis::viridis(256),
       row_title = "Group 4 subpopulations",
        column_title = "SHH subpopulations",
        cell_fun = function(j, i, x, y, width, height, fill) {
          bin <- findInterval(pvals[i, j], c(-log10(0.01), 20, 200))
          txt <- c("*", "**", "***")[bin]
          grid.text(txt, x, y, gp = gpar(fontsize = 10, col = "white"))
})


lgd_list <-  list(
    Legend(labels = c("*** < 1e-200",
                      "** < 1e-20",
                      "* < 1e-2"), 
           title = "Adj. p.val")
)
draw(h, heatmap_legend_list = lgd_list)


pdf(file.path(fig_dir, "Marker_overlap_gp4_shh.pdf"),
    width = 7,
    height = 7)
  draw(h, heatmap_legend_list = lgd_list)
dev.off()
```

## Other analyses

```{r, eval = FALSE}
all_tumor_mkrs <- wilcoxauc(so, "tumor_cell_type")

undiff_pops <- str_subset(unique(so$tumor_cell_type), "_Undifferentiated")
undiff_mkrs <- wilcoxauc(so, "tumor_cell_type", groups_use = undiff_pops)

top_undiff_mkrs <- undiff_mkrs %>% 
  filter(padj < 0.05,
                       logFC > 0, 
                       pct_in > 0.10) %>% 
  group_by(group) %>% 
  arrange(padj, desc(logFC), .by_group = T) %>% 
  ungroup()


diff_pops <- str_subset(unique(so$tumor_cell_type), "_Differentiated")
diff_mkrs <- wilcoxauc(so, "tumor_cell_type", groups_use = diff_pops)

top_diff_mkrs <- diff_mkrs %>% 
  filter(padj < 0.05,
                       logFC > 0, 
                       pct_in > 0.10) %>% 
  group_by(group) %>% 
  arrange(padj, desc(logFC), .by_group = T) %>% 
  ungroup()

top_diff_mkrs <- split(top_diff_mkrs, top_diff_mkrs$group) 
names(top_diff_mkrs) <- names(top_diff_mkrs) %>% 
   str_remove("_Differentiated") %>% 
  str_replace_all("[[:punct:]]", "")

write.xlsx(top_diff_mkrs, file.path(xcel_markers_dir,
                                    "tumor_differentiated_population_markers.xlsx")) 

top_undiff_mkrs <- split(top_undiff_mkrs, top_undiff_mkrs$group) 
names(top_undiff_mkrs) <- names(top_undiff_mkrs) %>% 
   str_remove("_Undifferentiated") %>% 
  str_replace_all("[[:punct:]]", "")

write.xlsx(top_undiff_mkrs, file.path(xcel_markers_dir,
                                    "tumor_undifferentiated_population_markers.xlsx")) 
```


```{r, eval = FALSE}
# Use top 500 markers   
mkrs_split <- map(top_undiff_mkrs, ~pull(.x[1:500, ], feature))

go_res <- gost(mkrs_split, 
               ordered_query = TRUE,
               sources = c("GO", 
                          "KEGG", 
                          "REAC", 
                          "TF"))

go_res[["result"]] %>% 
  mutate(parents = unlist(map(parents, ~str_c(.x, collapse = ","))),
         source = str_replace_all(source, "[[:punct:]]", " ")) %>% 
  select(-significant) %>% 
  split(., .$query) %>% 
  openxlsx::write.xlsx(., file.path(go_dir, "goterms_undifferentiated_cell_types.xlsx"))



mkrs_split <- map(top_diff_mkrs, ~pull(.x[1:500, ], feature))

go_res <- gost(mkrs_split, 
               ordered_query = TRUE,
               sources = c("GO", 
                          "KEGG", 
                          "REAC", 
                          "TF"))

go_res[["result"]] %>% 
  mutate(parents = unlist(map(parents, ~str_c(.x, collapse = ","))),
         source = str_replace_all(source, "[[:punct:]]", " ")) %>% 
  select(-significant) %>% 
  split(., .$query) %>% 
  openxlsx::write.xlsx(., file.path(go_dir, "goterms_differentiated_cell_types.xlsx"))
```

```{r, eval = FALSE}
mkrs <- diff_mkrs %>% 
  filter(padj < 0.05,
         logFC > 0, 
         pct_in > 0.10) %>% 
  group_by(group) %>% 
  arrange(padj, desc(logFC), .by_group = T) %>% 
  # mutate(group = factor(group, levels = ))
  slice(1:10) %>% 
  pull(feature)

tmp <- subset(so, 
              subset = tumor_cell_type %in% c(
                "SHH_Differentiated (GRIA2, GOLGA8A, MALAT1) #1",
                "SHH_Differentiated (STMN2/4, MLLT1, NHLH1/2) #4",
                "GP3_Differentiated (GOLGA8A, MALAT1) #0",
                "GP3_Differentiated Photoreceptor (NRL) #2",
                "GP4_Differentiated (GRIA2, GOLGA8A, MALAT1) #0",
                "GP4_Differentiated Photoreceptor (NRL) #5")
)

tmp@assays$TF <- NULL
tmp <- ScaleData(tmp, features = mkrs)

hmap <- plot_heatmap(tmp, 
             features = unique(mkrs),
             group = "tumor_cell_type",
             average = TRUE, 
             hmap_options = list(show_column_names = FALSE, 
                                 cluster_columns = TRUE,
                                 cluster_rows = FALSE)
    )

pdf(file.path(fig_dir, "top_markers_between_differentiated_populations.pdf"),
    width = 7,
    height = 9)
draw(hmap)
dev.off()

hmap



mkrs <- undiff_mkrs %>% 
  filter(padj < 0.05,
                       logFC > 0, 
                       pct_in > 0.10) %>% 
  group_by(group) %>% 
  arrange(padj, desc(logFC), .by_group = T) %>% 
 # mutate(group = factor(group, levels = ))
  slice(1:10) %>% 
  pull(feature)

tmp <- subset(so, 
              subset = tumor_cell_type %in% c(
                "SHH_Undifferentiated (RPGs and SFRP1) #0",
                "SHH_Undifferentiated (RPGs and BTF3) #2",
                "GP3_Undifferentiated (RPGs and BTF3) #1",
                "GP3_Undifferentiated (MYC, FOXQ1/F2) #4",
                "GP4_Undifferentiated (RPGs, STMN1/2, EOMES) #1",
                "GP4_Undifferentiated (RPGs, STMN1/2) #2")
)

tmp@assays$TF <- NULL
tmp <- ScaleData(tmp, features = mkrs)

hmap <- plot_heatmap(tmp, 
             features = unique(mkrs),
             group = "tumor_cell_type",
             average = TRUE, 
             hmap_options = list(show_column_names = FALSE, 
                                 cluster_columns = TRUE,
                                 cluster_rows = TRUE)
    )

pdf(file.path(fig_dir, "top_markers_between_progenitor_populations.pdf"),
    width = 7,
    height = 9)
draw(hmap)
dev.off()

hmap

```


### Cell cycle differences

```{r, eval = FALSE}
plts <- plot_feature(sos$GP4, c("tumor_cell_type", 
                        "Phase",
                        "MKI67",
                        "GRIA2",
                        "RPL32"),
             embedding = "pca")

p <- plot_grid(plotlist = plts[2:5],
  nrow = 2, 
  ncol = 2)

p

save_plot(file.path(fig_dir, "gp4_cell_cycle_pca.pdf"),
          p,
          nrow = 2,
          ncol = 2,
          base_asp = 1.5)

p <- plot_grid(plotlist = plts[1],
  nrow = 1, 
  ncol = 1)

p

save_plot(file.path(fig_dir, "gp4_cell_type_pca.pdf"),
          p,
          nrow = 1,
          ncol = 1,
          base_asp = 2.5)


p <- plot_harmony(sos$GP4, "tumor_cell_type", legend_title = "")

save_plot(file.path(fig_dir, "gp4_cell_type_umap.pdf"),
          p,
          base_asp = 2.5)
```


```{r, eval = FALSE}

plts <- plot_feature(sos$GP3, c("tumor_cell_type", 
                        "Phase",
                        "MKI67",
                        "PCDH9",
                        "RPL32"),
             embedding = "pca")

p <- plot_grid(plotlist = plts[2:5],
  nrow = 2, 
  ncol = 2)

p

save_plot(file.path(fig_dir, "gp3_cell_cycle_pca.pdf"),
          p,
          nrow = 2,
          ncol = 2,
          base_asp = 1.5)

p <- plot_grid(plotlist = plts[1],
  nrow = 1, 
  ncol = 1)

p

save_plot(file.path(fig_dir, "gp3_cell_type_pca.pdf"),
          p,
          nrow = 1,
          ncol = 1,
          base_asp = 2.5)

p <- plot_harmony(sos$GP3, "tumor_cell_type", legend_title = "")

save_plot(file.path(fig_dir, "gp3_cell_type_umap.pdf"),
          p,
          base_asp = 2.5)
```



```{r, eval = FALSE}

plts <- plot_feature(sos$SHH, c("tumor_cell_type", 
                        "Phase",
                        "MKI67",
                        "NHLH1",
                        "RPL32"),
             embedding = "pca")

p <- plot_grid(plotlist = plts[2:5],
  nrow = 2, 
  ncol = 2)

p

save_plot(file.path(fig_dir, "shh_cell_cycle_pca.pdf"),
          p,
          nrow = 2,
          ncol = 2,
          base_asp = 1.5)

p <- plot_grid(plotlist = plts[1],
  nrow = 1, 
  ncol = 1)

p

save_plot(file.path(fig_dir, "shh_cell_type_pca.pdf"),
          p,
          nrow = 1,
          ncol = 1,
          base_asp = 2.5)


p <- plot_harmony(sos$SHH, "tumor_cell_type", legend_title = "")

save_plot(file.path(fig_dir, "shh_cell_type_umap.pdf"),
          p,
          base_asp = 2.5)
```


## Summary files/heatmaps

```{r}

so_fns <- file.path(obj_dir, c(
  file.path("shh", "shh.qs"),
  file.path("gp34", "gp3_only.qs"),
  file.path("gp34", "gp4_only.qs")))
                    
clusters <- c(
  "tumor_subpopulations_shh",
  "tumor_subpopulations_gp3",
  "tumor_subpopulations_gp4"
)

sos <- map(so_fns, 
            ~qread(.x, nthreads = 4, use_alt_rep = FALSE))

names(sos) <- c("SHH", "GP3", "GP4")

sos <- pmap(
  list(sos,
    names(sos),
    clusters),
            function(so, id, cluster){
              # drop the excluded populations
              to_keep <- rownames(so@meta.data)[!str_detect(so[[cluster]][,1], "-X[0-9]$")]
              so <- subset(so, cells = to_keep)
              so$tumor_cell_type <- so[[cluster]][,1]
              so
            })
```

### TF data
 
Make heatmaps of top TFs per subpopulation

```{r}
iwalk(sos,
     function(so, id){
       DefaultAssay(so) <- "TF"
       uid <- id
       lid <- str_to_lower(id)
       so_tf_markers <- read_tsv(file.path("tables",
                                           lid,
                                           str_c("subpopulation_tf_activities_", uid, ".tsv")))
       
       topx <- so_tf_markers %>% 
         group_by(group) %>% 
         arrange(padj, desc(logFC), .by_group = TRUE) %>% 
         slice(1:5)
       
       so <- ScaleData(so)
       
       p <- DoHeatmap(so, 
                      group.colors = discrete_palette_default,
                      features = unique(topx$feature),
                      group.by = str_c("tumor_subpopulations_", lid),
                      angle = 45,
                      raster = FALSE, 
                      draw.lines = TRUE)
       
       p <- p +
         scale_fill_gradientn(colours = viridis::viridis(256),
                              name = "Regulon AUC\nZ-scores")
       
       save_plot(file.path(fig_dir, str_c("heatmap_tfs_",
                                 lid, 
                                 ".pdf")), 
                           p,
                           base_asp = 2,
                           base_height = 8)
     })

```

Make single spreadsheet of TF markers per subpopulation

```{r}
dir.create(file.path("tables", "subpopulation_summaries"), showWarnings = FALSE)
sgs <- c("SHH", "GP3", "GP4")
all_tumor_tf_dat  <- map_dfr(sgs, 
    ~read_tsv(file.path("tables", 
                        .x,  
                        str_c("subpopulation_tf_activities_", .x, ".tsv")))) %>%
      split(., .$group)

write_markers_xlsx(all_tumor_tf_dat,  
                   path = file.path(tbls_dir, "subpopulation_summaries","subpopulation_TF_activities.xlsx"),
                   description_string = "Transcription factor regulons derived from PyScenic that are significantly enriched in each tumor subpopulation")
                   
```


### Go-term data
 
Make heatmaps of top Goterms per subpopulation

```{r}
iwalk(sos,
     function(so, id){
       
       so_markers <- read_tsv(file.path("markers",
                                        id, 
                                        str_c("harmony_subpopulation_markers_", id, ".tsv"))) %>% 
         filter(!str_detect(group, "-X[0-9]$"))
       
       # Go terms using top 500 markers 
       mkrs <- so_markers %>% 
         group_by(group) %>% 
         slice(1:500)  
    
    mkrs_split <- split(mkrs, mkrs$group) %>% 
      map(~pull(.x, feature))
    
    go_res <- gost(mkrs_split, 
                   ordered_query = TRUE,
                   significant = FALSE,
                   sources = c("GO", 
                               "KEGG", 
                               "REAC", 
                               "TF"))
    
    saveRDS(go_res, file.path(go_dir, str_c(id, "_all_go_terms.rds")))
    
    walk(c("GO:BP", "KEGG", "REAC"),
         ~{
           go_mat <- go_res$result %>% 
             filter(source == .x) %>% 
             dplyr::select(query, p_value, term_name) %>% 
             pivot_wider(names_from = query, values_from = p_value) %>% 
             column_to_rownames("term_name") %>% 
             as.matrix() 
           # set terms that are NA to 1
           go_mat[is.na(go_mat)] <- 1
           go_mat <- -log10(go_mat)
           
           top_terms <- go_res$result %>% 
             filter(source == .x) %>% 
             group_by(query) %>% 
             arrange(p_value, .by_group = TRUE) %>% 
             slice(1:5) %>% 
             pull(term_name) %>% 
             unique()
           row_go_labels <- rownames(go_mat[top_terms, ]) %>% 
             str_split(" ") %>% 
             map_chr(~str_c(.x[1:min(length(.x), 10)], collapse = " ")) %>% 
             str_wrap(, 50)
           hmap <- Heatmap(go_mat[top_terms, ], 
                           name = "-log(10) pvalue", 
                           col = viridis::viridis(256),
                           column_order = unique(go_res$result$query),
                           cluster_rows= FALSE,
                           row_names_gp = gpar(fontsize = 12),
                           row_names_side = "left",
                           row_labels = row_go_labels,
                           row_names_max_width = max_text_width(row_go_labels, 
                                                                gp = gpar(fontsize = 12)),
                           column_names_side = "top")
           pdf(file.path("figs", id, 
                         str_c("heatmap_goterms_", 
                               str_replace(.x, ":", "_"),
                               "_",
                               id, 
                               ".pdf")),
               width = 9,
               height = 7)
             draw(hmap)
           null <- dev.off()
         })
   
     })

```

Make single spreadsheet of TF markers per subpopulation

```{r}
map_dfr(names(sos), ~readRDS(file.path(go_dir, str_c(.x, "_all_go_terms.rds"))) %>% 
      .$result %>% 
      filter(p_value < 0.05) %>% 
  dplyr::select(-significant)) %>% 
  split(., .$query) %>% 
  openxlsx::write.xlsx(., 
                       file.path("tables", "subpopulation_summaries","subpopulation_goterm_enrichments.xlsx"))
                   
```

Make heatmap highlighting neuronal differentation related pathways

```{r}
go_res <- map(names(sos), ~readRDS(file.path(go_dir, str_c(.x, "_all_go_terms.rds"))))
names(go_res) <- names(sos)

# top terms containing "neuron" or "nervous" from SHH-C2 
to_plot <- c("nervous system development",
             "neuron development",
             "neuron differentiation",
             "generation of neurons",
             "neuron projection development",
             "neuron projection morphogenesis",
             "cell morphogenesis involved in neuron differentiation",
             "regulation of nervous system development",
             "regulation of neuron differentiation",
             "regulation of neuron projection development")



hmaps <- imap(go_res, ~{
  go_mat <- .x$result %>% 
    filter(source == "GO:BP",
           term_name %in% to_plot) %>% 
    dplyr::select(query, p_value, term_name) %>% 
    pivot_wider(names_from = query, values_from = p_value) %>% 
    column_to_rownames("term_name") %>% 
    as.matrix() 
  # set terms that are NA to 1
  go_mat[is.na(go_mat)] <- 1
  go_mat <- -log10(go_mat)
  
  
  row_go_labels <- rownames(go_mat) %>% 
    str_split(" ") %>% 
    map_chr(~str_c(.x[1:min(length(.x), 10)], collapse = " ")) %>% 
    str_wrap(, 50)
  hmap <- Heatmap(go_mat, 
                  name = "-log(10) pvalue", 
                  col = viridis::viridis(256),
                  column_order = unique(.x$result$query),
                  cluster_rows = TRUE,
                  show_row_dend = FALSE,
                  row_names_gp = gpar(fontsize = 12),
                  row_names_side = "left",
                  row_labels = row_go_labels,
                  row_names_max_width = max_text_width(row_go_labels, 
                                                       gp = gpar(fontsize = 12)),
                  column_names_side = "top")
  pdf(file.path("figs", .y, 
                str_c("heatmap_neuron_development_goterms_", 
                      .y, 
                      ".pdf")),
      width = 8,
      height = 5)
  draw(hmap)
  null <- dev.off()
  
  hmap
})

```
