---
title: "Data preprocessing"
author: "Kent Riemondy RBI"
date: "`r Sys.Date()`"
knit: (function(inputFile, encoding) {
  rmarkdown::render(inputFile, encoding = encoding, output_dir = "html") })
output:
  html_document:
    toc: true
    toc_float: true
    toc_collapsed: false
    theme: cosmo
    highlight: tango
    fig_caption: true
    code_folding: hide
    df_print: paged
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_chunk$set(cache = FALSE)
knitr::opts_chunk$set(warning = FALSE)
knitr::opts_chunk$set(cache.lazy = FALSE)
```


```{r libs, message=FALSE, warning=FALSE, echo=FALSE}
source(here::here("/R/utils.R"))

rmd_dir <- "preprocess"

fig_dir <- file.path(fig_dir, rmd_dir)
mkrs_dir <- file.path(mkrs_dir, rmd_dir)
tbls_dir <- file.path(tbls_dir, rmd_dir)
obj_dir <- file.path(obj_dir, rmd_dir)

walk(c(fig_dir, mkrs_dir, tbls_dir, obj_dir),
     dir.create, showWarnings = F)

seed_value <- 20200204
```

# Experiment Summary

```{r get_data}

# read in metadata to parse out directories and sample ids
mdata <- read_excel(file.path(doc_dir, 
                                   "MED scRNAseq batch and annotations 5.8.19.xlsx")) %>% 
  filter(subgroup != "LGG") %>% 
  select(-subgroup)

mdata_updated <- read_excel(file.path(doc_dir, 
                                   "MED scRNAseq batch and annotations no TBD 5.28.19.xlsx")) %>% 
  rename(subgroup = `subgroup by transcriptomics` , 
         date = ...7,
         fq_id = ...8 ) %>% 
  filter(subgroup != "LGG") %>% 
  select(UPN, subgroup)

mdata <- left_join(mdata_updated, mdata, by = "UPN")


sample_paths <- file.path(data_dir,
                          mdata$fq_id,
                          "outs", 
                          "filtered_feature_bc_matrix")

names(sample_paths) <- str_replace(mdata$UPN, "_", "-")
```

```{r}
mat <- Read10X(sample_paths)
```

## General QC for library prep  {.tabset}

```{r }
metrics_paths <- file.path(data_dir,
                          mdata$fq_id,
                          "outs", 
                          "metrics_summary.csv")

names(metrics_paths) <- mdata$UPN

mapping_dat <- map_dfr(metrics_paths, read_csv, .id = "sample")

clean_up_metadata <- function(metrics_summary) {
  metrics_summary <- mutate_all(metrics_summary, str_remove, "%$")
  metrics_summary <- mutate_at(metrics_summary, .vars= 2:ncol(metrics_summary), as.numeric)
  metrics_summary
}

mapping_dat <- clean_up_metadata(mapping_dat)

metrics <- colnames(mapping_dat)[2:ncol(mapping_dat)]

mapping_dat <- mapping_dat %>% 
  gather(metric, value, -sample) %>% 
  left_join(mdata, by = c("sample" = "UPN")) %>% 
  arrange(subgroup) %>% 
  mutate(sample = factor(sample, levels = unique(sample)))

p <- map(metrics, 
    function(x) {
    filter(mapping_dat, metric == x) %>% 
          ggplot(aes(sample, value)) +
            geom_point(aes(color = subgroup)) +
        scale_color_brewer(palette = "Paired") + 
        labs(y = x, 
             x = "") +
        theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5))
}) 
```

```{r, results ='asis'}

for(i in seq_along(p)){
  cat('\n### ', metrics[i], '\n')
  print(p[i])
  cat('\n')
}
```

## Preprocessing {.tabset}

## Rerun preprocessing scaling, PCA, UMAP, etc. 

```{r create_seurat, message = F, results = 'hide', warning = F}
so <- CreateSeuratObject(
  mat,
  min.cells = 3,
  min.features = 200,
  names.delim = "_",
  names.field = 1
)

rm(mat)
gc()

# fix 966-2 to 966_2 
so@meta.data$orig.ident <- str_replace(so@meta.data$orig.ident,
                                         "966-2", 
                                         "966_2")

so <- PercentageFeatureSet(so, 
                             pattern = "^MT-", 
                             col.name = "percent.mt")

so@meta.data <- get_metadata(so, embedding = NULL) %>% 
  left_join(mdata, by = c("orig.ident" = "UPN")) %>% 
  mutate(UPN = orig.ident) %>% 
  as.data.frame() %>% 
  tibble::column_to_rownames("cell")

# Fix 934 and 943, which are the same
# and annotate that 966_2 is a recurrance
so@meta.data <- get_metadata(so, embedding = NULL) %>% 
  mutate(additional_info = case_when(
    UPN == "943" ~ "repeated_sample",
    UPN == "966_2" ~ "recurrance",
    UPN == "966" ~ "primary",
    TRUE ~ NA_character_),
         UPN = ifelse(UPN == "943",
                      "934",
                      UPN),
         UPN = ifelse(UPN == "966_2",
                      "966",
                      UPN)) %>% 
  as.data.frame() %>% 
  tibble::column_to_rownames("cell")

saveRDS(so, file.path(obj_dir, "unfiltered.rds"))
```

### Percent Mitochondrial UMIs 

```{r}
p <- plot_violin(so@meta.data, 
            "orig.ident",
            "percent.mt",
            .fill = "subgroup") +
  labs(x = "", y = "Percent UMIs from Mitochondria") + 
 geom_hline(aes(yintercept = 30), linetype = "dashed")

save_plot(file.path(fig_dir, "prop_mito.pdf"),
          p, base_asp = 1.75)
p
```


### # of genes detected 

```{r}
p <- plot_violin(so@meta.data,
            "orig.ident", 
            "nFeature_RNA",
            .fill = "subgroup") +
  labs(x = "", y = "# of genes per cell")

save_plot(file.path(fig_dir, "ngene.pdf"),
          p, base_asp = 1.75)
p
```


### # of UMIs detected

```{r}
p <- plot_violin(so@meta.data, 
            "orig.ident",
            "nCount_RNA", 
            .fill = "subgroup") +
  labs(x = "", y = "# of UMIs") + 
 geom_hline(aes(yintercept = 40000), linetype = "dashed")

save_plot(file.path(fig_dir, "numi.pdf"),
          p, base_asp = 1.75)
p
```

### Table of mitochondrial proportions per sample

```{r}
so@meta.data %>% 
  group_by(UPN) %>% 
  summarize(median_percent_mito = median(percent.mt), 
            mean_percent_mito = mean(percent.mt)) %>% 
  arrange(desc(median_percent_mito))
```


## Relationship between UMIs and % mitochondria {.tabset}

### All samples
```{r, fig.width = 24, fig.height = 18}
sample_names <- as.character(unique(so@meta.data$orig.ident))
per_sample <- map(sample_names, ~filter(so@meta.data, 
                                        orig.ident == .x))
p <- list()
for(i in seq_along(per_sample)){
  .col <- discrete_palette_default[i]
  p[[i]] <- ggplot(per_sample[[i]], aes(nCount_RNA, percent.mt)) +
        geom_point(aes(color = UPN)) +
        scale_color_manual(values = .col)
}

plt <- plot_grid(plotlist= p, base_asp = 1, nrow = 6, ncol = 6)


save_plot(file.path(fig_dir, "umi_v_prop_mito.pdf"),
          plt,
          nrow = 6,
          ncol = 6)

plt
```

```{r, results ='asis'}

for(i in seq_along(per_sample)){
  .col <- discrete_palette_default[i]
  cat('\n### ', sample_names[i], '\n')
  p <- ggplot(per_sample[[i]], aes(nCount_RNA, percent.mt)) +
        geom_point(aes(color = UPN)) +
        scale_color_manual(values = .col)
  print(p)
  cat('\n')
}
```


## Filter cells and samples.

Suggest removing cells with > 30% mitochondrial UMIs, > 40K UMIs. 

```{r}
so <- readRDS(file.path(obj_dir, "unfiltered.rds"))

so <- subset(so, subset = percent.mt < 30)
so <- subset(so, subset = nCount_RNA < 40000)
so <- subset(so, subset = nCount_RNA > 500)
```


Annotate samples and assign colors to each sample. 

```{r}
annots <- so@meta.data %>% 
  select(UPN, subgroup) %>% 
  unique() %>%
  arrange(subgroup)

subgroup_order <- c(
  "SHH",
  "GP3",
  "GP3/4",
  "GP4",
  "WNT"
)

annots <- mutate(annots, subgroup = factor(subgroup, levels = subgroup_order))

annots <- annots %>% arrange(subgroup, as.numeric(UPN))

sample_order <- annots$UPN 

so@meta.data$UPN <- factor(so@meta.data$UPN, levels = sample_order)
so@meta.data$subgroup <- factor(so@meta.data$subgroup, levels = subgroup_order)
```


## Normalize and embed into 2D with UMAP

```{r}
so <- NormalizeData(so)

so <- FindVariableFeatures(
  so,
  selection.method = "vst",
  nfeatures = 4000,
  verbose = FALSE
)

so <- ScaleData(so, verbose = TRUE)

so <- RunPCA(so, 
             npcs = 40, 
             verbose = FALSE, 
             seed.use = seed_value)

ElbowPlot(so, ndims = 50)

# make graphs and use graphs for UMAP
so <- FindNeighbors(so, 
                    reduction = "pca", 
                    dims = 1:40, 
                    k.param = 20L)

so <- RunUMAP(so, 
              graph = "RNA_snn", 
              min.dist = 0.4, 
              seed.use = seed_value)

so <- FindClusters(so, 
                   resolution = c(0.1, 0.2, 0.3, 0.5),
                   random.seed = seed_value)

so$coarse_clusters <- so$RNA_snn_res.0.1

Idents(so) <- "coarse_clusters"
```

```{r}
plot_umap(so, "subgroup", 
          .col = palette_OkabeIto, 
          label_text = TRUE, 
          label_col = "black",
          sorted = "random")
```

```{r}
plot_umap(so, "coarse_clusters", label_text = TRUE, label_col = "black")
plot_umap(so, "UPN", 
          label_text = TRUE, 
          label_col = "black", 
          sorted = "random")
```

## markers 

```{r}
mkrs <- wilcoxauc(so, "coarse_clusters")

mkrs <- filter(mkrs, 
       logFC > 0, padj < 0.05) %>% 
  group_by(group) %>% 
  arrange(padj, desc(logFC)) %>% 
  ungroup()

write_tsv(mkrs, file.path(mkrs_dir, "coarse_cluster_markers_all_data_unfiltered.tsv"))

mkrs %>% 
  split(., .$group) %>% 
  write_markers_xlsx(.,
                   file.path(tbls_dir, "coarse_cluster_markers_all_data_unfiltered.xlsx"))
```

```{r, fig.width = 20, fig.height= 20}
topx <- mkrs %>% 
  filter(logFC > 0, padj < 0.05) %>% 
  mutate(group = as.factor(as.integer(group))) %>% 
  group_by(group) %>% 
  arrange(padj, desc(logFC)) %>% 
  slice(1:10)

p <- DotPlot(so, 
        features = unique(topx$feature),
        group.by = "coarse_clusters") +
  coord_flip() +
  labs(x = "Cluster",
       y = "")

p

save_plot(file.path(fig_dir, 
                    "coarse_cluster_marker_dotplot_unfiltered.pdf"), 
          p, base_height = 24, base_asp = 1)
```


```{r}
so <- calc_diversity(so, "UPN", "coarse_clusters")

so$cdiversity <- so$entropy
so$entropy <- NULL

p <- plot_umap(so, "cdiversity", legend_title = "cluster\ndiversity")

save_plot(file.path(fig_dir, "cluster_diversity_unfiltered.pdf"), p, base_asp = 1.2)

p
```

```{r}
plot_umap(so, "coarse_clusters", label_text = TRUE)
```

```{r}
plot_umap(so, "UPN")
```


## Annotate coarse cell types

```{r}
cids <- as.numeric(as.character(so$coarse_clusters)) %>% unique() %>% sort()

new_ids <- c(
  "13" = "lymphocytes",
  "1" = "macrophage_monocytes",
  "15" ="oligodendrocytes_astrocytes_other",
  "5" = "lymphocytes"
)
others <- setdiff(cids, names(new_ids))
other_ids <- rep("malignant", length(others))
names(other_ids) <- as.character(others)

new_ids <- c(new_ids, other_ids)

so$coarse_cell_type <- new_ids[as.character(so$coarse_clusters)]

plot_umap(so, "coarse_cell_type", .cols = palette_OkabeIto)
```



```{r}

to_plot <- c(
  "UPN",
  "coarse_clusters",
  "coarse_cell_type",
  "subgroup",
  "cdiversity"
) 

pals <- list(
  NULL,
  NULL,
  palette_OkabeIto,
  palette_OkabeIto,
  NULL
)

plts <- map2(to_plot, pals,
    ~plot_umap(so, .x, .cols = .y))

plts
```


```{r}
p <- plot_cell_proportions(so, 
                      group_id = "coarse_cell_type",
                      facet_by = "subgroup", 
                      cols = rev(palette_OkabeIto[1:4])) +
  labs(fill = "", x = "")

p

save_plot(file.path(fig_dir, "cell_proportion_no_filtering.pdf"), 
          p, 
          base_asp = 2)
```

## refilter to exclude poor samples

Removing sample 1008 as it only has ~45 malignant cells and is dominated by immune cells, same with 1192 from the WNTs. 1202 appears to be a sample swap as it has a lot of SHH character and inferCNV results do not match the methylation data. Excluding

```{r}
to_drop <- c("1008", "1192", "1202")
to_keep <- colnames(so)[which(!so@meta.data$orig.ident %in% to_drop)]
so <- subset(so, cells = to_keep)
```


Annotate samples and assign colors to each sample. 

```{r}
annots <- so@meta.data %>% 
  select(UPN, subgroup) %>% 
  unique() %>%
  arrange(subgroup)

subgroup_order <- c(
  "SHH",
  "GP3",
  "GP3/4",
  "GP4",
  "WNT"
)

annots <- mutate(annots, subgroup = factor(subgroup, levels = subgroup_order))

annots <- annots %>% arrange(subgroup, as.numeric(UPN))

sample_order <- annots$UPN 

so@meta.data$UPN <- factor(so@meta.data$UPN, levels = sample_order)
so@meta.data$subgroup <- factor(so@meta.data$subgroup, levels = subgroup_order)
```


## Normalize and reembed into 2D with UMAP

```{r}
so <- NormalizeData(so)

so <- FindVariableFeatures(
  so,
  selection.method = "vst",
  nfeatures = 4000,
  verbose = FALSE
)

so <- ScaleData(so, 
                vars.to.regress = "percent.mt", 
                verbose = TRUE)

so <- RunPCA(so, 
             npcs = 40, 
             verbose = FALSE, 
             seed.use = seed_value)

ElbowPlot(so, ndims = 50)

# make graphs and use graphs for UMAP
so <- FindNeighbors(so, 
                    reduction = "pca", 
                    dims = 1:40, 
                    k.param = 20L)

so <- RunUMAP(so, 
              graph = "RNA_snn", 
              min.dist = 0.4, 
              seed.use = seed_value)

so <- FindClusters(so, 
                   resolution = c(0.1, 0.2, 0.3, 0.5),
                   random.seed = seed_value)

so$coarse_clusters <- so$RNA_snn_res.0.1

Idents(so) <- "coarse_clusters"
```

```{r}
plot_umap(so, 
          "subgroup", 
          .col = palette_OkabeIto, 
          label_text = TRUE, 
          label_col = "black",
          sorted = "random")
```

```{r}
plot_umap(so, "coarse_clusters", 
          label_text = TRUE, label_col = "black")
plot_umap(so, "UPN", 
          label_text = TRUE, 
          label_col = "black", 
          sorted = "random")
```

## markers 

```{r}
mkrs <- wilcoxauc(so, "coarse_clusters")

mkrs <- filter(mkrs, 
       logFC > 0, padj < 0.05) %>% 
  group_by(group) %>% 
  arrange(padj, desc(logFC)) %>% 
  ungroup()

write_tsv(mkrs, file.path(mkrs_dir, "coarse_cluster_markers_all_data.tsv"))

mkrs %>% 
  split(., .$group) %>% 
  write_markers_xlsx(.,
                   file.path(tbls_dir, "coarse_cluster_markers_all_data.xlsx"))
```

```{r, fig.width = 20, fig.height= 20}
topx <- mkrs %>% 
  filter(logFC > 0, padj < 0.05) %>% 
  mutate(group = as.factor(as.integer(group))) %>% 
  group_by(group) %>% 
  arrange(padj, desc(logFC)) %>% 
  slice(1:10)

p <- DotPlot(so, 
        features = unique(topx$feature),
        group.by = "coarse_clusters") +
  coord_flip() +
  labs(x = "Cluster",
       y = "")

p

save_plot(file.path(fig_dir, 
                    "coarse_cluster_marker_dotplot.pdf"), 
          p, base_height = 24, base_asp = 1)
```


```{r}
so <- calc_diversity(so, "UPN", "coarse_clusters")

so$cdiversity <- so$entropy
so$entropy <- NULL

p <- plot_umap(so, "cdiversity", legend_title = "cluster\ndiversity")

save_plot(file.path(fig_dir, "cluster_diversity.pdf"), p, base_asp = 1.2)

p
```

```{r}
plot_umap(so, "coarse_clusters", label_text = TRUE)
```

```{r}
plot_umap(so, "UPN")
```


## Annotate coarse cell types

```{r}
cids <- as.numeric(as.character(so$coarse_clusters)) %>% unique() %>% sort()

new_ids <- c(
  "11" = "lymphocytes",
  "3" = "macrophage_monocytes",
  "15" ="oligodendrocytes_astrocytes_other"
)
others <- setdiff(cids, names(new_ids))
other_ids <- rep("malignant", length(others))
names(other_ids) <- as.character(others)

new_ids <- c(new_ids, other_ids)

so$coarse_cell_type <- new_ids[as.character(so$coarse_clusters)]

plot_umap(so, "coarse_cell_type", .cols = palette_OkabeIto)
```



```{r}
to_plot <- c(
  "UPN",
  "coarse_clusters",
  "coarse_cell_type",
  "subgroup",
  "cdiversity"
) 

pals <- list(
  NULL,
  NULL,
  palette_OkabeIto,
  palette_OkabeIto,
  NULL
)

plts <- map2(to_plot, pals,
    ~plot_umap(so, .x, .cols = .y))

plts
```


```{r}
p <- plot_cell_proportions(so, 
                      group_id = "coarse_cell_type",
                      facet_by = "subgroup", 
                      cols = rev(palette_OkabeIto[1:4])) +
  labs(fill = "", x = "")

p

save_plot(file.path(fig_dir, "cell_proportions.pdf"), 
          p, 
          base_asp = 2)
```

## Write out data

```{r} 
qsave(so, file.path(obj_dir, "so.qs"), nthreads = 4)

mdata_out <- get_metadata(so)
outname <- file.path("tables", 
                     paste0("metadata_", 
                            format(Sys.Date(), "%Y_%m_%d"),
                            ".tsv.gz"))
write_tsv(mdata_out, outname)

```

## Get top PC genes

```{r}
so <- qread(file.path(obj_dir, "so.qs"), nthreads = 4)

to_plot <- c("UPN", "coarse_cell_type", "subgroup")

dims_to_plot <- list(
  c(1,2),
  c(2,3),
  c(3,4),
  c(4,5))
  
plts <- list()
for ( i in seq_along(dims_to_plot)){
  plts[[i]] <- map(to_plot, ~plot_pca(so, .x, dims =dims_to_plot[[i]])) %>%
  plot_grid(plotlist = ., nrow = 1, ncol = length(to_plot),
            rel_widths = c(1.75, 1.5, 1))
}

names(plts) <- str_c("pc_", 
                     map_chr(dims_to_plot, ~str_c(.x, collapse = "_")),
                     ".pdf")
pc_dir <- file.path(fig_dir, "pcs_all_cells")
dir.create(pc_dir)
iwalk(plts,
  ~save_plot(file.path(pc_dir, .y), .x, nrow = 1, ncol = 3))
```
