---
title: "Group 3 mouse model Myc + Gfi1"
author: "Kent Riemondy RBI"
date: "`r Sys.Date()`"
knit: (function(inputFile, encoding) {
  rmarkdown::render(inputFile, encoding = encoding, output_dir = "html") })
output:
  html_document:
    toc: true
    toc_float: true
    toc_collapsed: false
    theme: cosmo
    highlight: tango
    fig_caption: true
    code_folding: hide
    df_print: paged
---

```{r libs, message=FALSE, warning=FALSE, echo=FALSE}
source(here::here("/R/utils.R"))

rmd_dir <- file.path("mouse", "gp34")

fig_dir <- file.path(fig_dir, rmd_dir)
mkrs_dir <- file.path(mkrs_dir, rmd_dir)
tbls_dir <- file.path(tbls_dir, rmd_dir)
obj_dir <- file.path(obj_dir, rmd_dir)

walk(c(fig_dir, mkrs_dir, tbls_dir, obj_dir),
     dir.create, showWarnings = F)


seed_value <- 20200304
library(igraph)
```

## Get ortholog table

Download list of orthologs between mouse and human as gene symbols.

```{r}
ortholog_table <- "../dbases/mouse_orthologs.tsv"

if(!file.exists(ortholog_table)){
  library(biomaRt)
  ensembl <- useMart("ensembl", dataset = "hsapiens_gene_ensembl")
  
  orthologs <- getBM(mart = ensembl, 
      attributes = c("external_gene_name",
                     "mmusculus_homolog_associated_gene_name"))
  
  write_tsv(orthologs, ortholog_table)
}

orthologs <- read_tsv(ortholog_table)

orthologs <- filter(orthologs,
                            !is.na(mmusculus_homolog_associated_gene_name)) %>% 
  group_by(external_gene_name) %>% 
  mutate(n_orthos_h = n()) %>% 
  group_by(mmusculus_homolog_associated_gene_name) %>% 
  mutate(n_orthos_m = n()) %>% 
  filter(n_orthos_h == 1, n_orthos_m == 1) %>% 
  select(-starts_with("n_orthos")) %>% 
  ungroup()

orthologs <- mutate(orthologs, 
                    ortho_id = str_c(external_gene_name, ":",
                                     mmusculus_homolog_associated_gene_name))

cc_mouse_genes <- map(cc.genes.updated.2019, 
                      ~tibble(external_gene_name = .x) %>% 
                        left_join(orthologs, by = "external_gene_name") %>% 
                        na.omit() %>% 
                        pull(mmusculus_homolog_associated_gene_name))
```

>
For the mouse samples
MP control and MP luc-GFP are similar and both from MP (MYC + DNP53 expression in CD133+ cells) tumor models expressing luciferase. (DN= dominant negative p53)
MG CD2 luc is the second model (MYC + GFI1 expression in CD133+ cells).
>

10.1016/j.ccr.2011.12.021


## Samples

```{r get_data}
samples <- c(
  "1_Mpcont",
  "2_MP_luc_GFP",
  "3_MG"
)

sample_paths <- file.path(data_dir,
                          samples,
                          "outs", 
                          "filtered_feature_bc_matrix")

names(sample_paths) <- samples

mat <- Read10X(sample_paths)

so <- CreateSeuratObject(
  mat,
  min.cells = 3,
  min.features = 200,
  names.delim = "_",
  names.field = c(1, 2)
)

so@meta.data <- so@meta.data %>% 
  tibble::rownames_to_column("cell") %>% 
  mutate(UPN = orig.ident) %>% 
  as.data.frame() %>% 
  tibble::column_to_rownames("cell")

# add in new sample label
new_expt_labels <- c(
  "2_MP" = "MYC + DNP53",
  "3_MG" = "MYC + GFI1"
)

so@meta.data$expt_ids <- new_expt_labels[as.character(so@meta.data$orig.ident)]

rm(mat)
a <- gc()
```


# MYC + DNP53 

```{r}
# new_expt_labels <- c(
#   "2_MP" = "MYC + DNP53",
#   "3_MG" = "MYC + GFI1"
# )
seed_value <- 20200301
res_settings <- c(0.1, 0.2, 0.3, 0.4, 0.5, 0.6)
so <- subset(so, subset = orig.ident == "3_MG") %>% 
  NormalizeData(verbose = FALSE) %>% 
  FindVariableFeatures(selection.method = "vst", 
                       nfeatures = 3000, 
                       verbose = FALSE) %>% 
  ScaleData(verbose = FALSE) %>% 
  RunPCA(npcs = 40, 
         verbose = FALSE, 
         seed.use = seed_value) %>% 
  FindNeighbors(reduction = "pca", 
                dims = 1:30,
                verbose = FALSE) %>% 
  RunUMAP(dims = 1:30,
          min.dist = 0.3,
          seed.use = seed_value,
          verbose = FALSE) %>% 
  FindClusters(resolution = res_settings,
               verbose = FALSE) %>% 
  CellCycleScoring(s.features = cc_mouse_genes$s.genes,
                   g2m.features = cc_mouse_genes$g2m.genes) %>% 
  PercentageFeatureSet(pattern = "^mt-", 
                       col.name = "percent.mt")

so$seurat_clusters <- NULL
```

### Annotate with gfp/luc reads

```{r count_gfp}
count_transgenes <- function(sam_path, cb_prefix = "") {
  sam_lines <- read_lines(sam_path) %>%
    .[!str_detect(., "^@")] %>%
    str_split(., "\t")
  
  
  alignment_lines <- map_dfr(sam_lines,
                             function(x) {
                               read_info <- str_remove(x[1], "_$")
                               read_info <-
                                 str_split(read_info, "_")[[1]]
                               
                               if (length(read_info) == 3) {
                                 read <- read_info[1]
                                 cb <- read_info[2]
                                 umi <- read_info[3]
                               } else{
                                 read <- read_info[1]
                                 cb <- NA
                                 umi <- NA
                               }
                               
                               contig <- x[3]
                               flag <- x[2]
                               tibble(read, cb, umi, contig, flag)
                             }) %>%
    na.omit() %>%
    filter(flag == 0)
  
  
  res <- group_by(alignment_lines, cb, umi, contig) %>%
    summarize(n()) %>%
    group_by(cb, contig) %>%
    summarize(n_umi = n()) %>% 
    ungroup()
  # fix cell barcode ids
  
  res <- mutate(res, 
                cb = str_remove(cb, "^CB:Z:"),
                cb = str_remove(cb, "-[0-9]$"),
                cb = str_c(cb_prefix, cb))
  
  res <- spread(res, contig, n_umi, fill = 0L)
  res
}

sample_paths <- file.path(data_dir,
                          samples,
                          "outs",
                          "tg.sam")

tg_counts <- map2_dfr(sample_paths, str_c(samples, "_"), count_transgenes) %>% 
  mutate_if(is.numeric, log1p)

so@meta.data <- so@meta.data %>% 
  rownames_to_column("cb") %>% 
  left_join(tg_counts, by = "cb") %>%
  mutate(egfp = ifelse(is.na(egfp),
                       0L,
                       egfp),
         `luc-firefly` = ifelse(is.na(`luc-firefly`),
                       0L,
                       `luc-firefly`)) %>% 
  column_to_rownames("cb") 

to_plot <- c("egfp", "luc-firefly")
names(to_plot) <- c("egfp", "luc-firefly")

plts <- imap(to_plot, 
    ~plot_umap(so, .x))  

plts
iwalk(plts, 
    ~save_plot(file.path(fig_dir, 
                         paste0("umap_gp34_myc_gfi1_by_", .y, ".pdf")),
          .x, 
          base_asp = 1.2))

```



```{r}
cluster_cols <- str_subset(colnames(so@meta.data), "res")

plts <- plot_umap(so, cluster_cols) 
names(plts) <- cluster_cols

iwalk(plts, 
    ~save_plot(file.path(fig_dir, 
                         paste0("umap_gp34_gfi1_by_", .y, ".pdf")),
          .x, 
          base_asp = 1))

plt <- plot_grid(plotlist = plts,
                 nrow = 2, 
                 ncol = 3)

save_plot(file.path(fig_dir, 
                    paste0("umap_gp34_gfi1_all_clusters.pdf")),
          plt,
          nrow = 2, 
          ncol = 3, 
          base_asp = 1.5)
plt
```


```{r}
so$clusters <- so$RNA_snn_res.0.3

library(presto)

mkrs <- get_marker_summaries(so, 
                             "clusters",
                             prefix = "myc_gfi1",
                     tsv_output_dir = mkrs_dir,
                     xlsx_output_dir = tbls_dir)
mkrs
```


```{r}
top_features <- mkrs$mkrs %>%  
  ungroup() %>% 
  mutate(group = as.factor(as.integer(group))) %>% 
  group_by(group) %>% 
  dplyr::slice(1:10) %>% 
  pull(feature) %>% 
  unique()

tmp_so <- so[top_features, ] %>% 
  ScaleData(features = top_features)

p <- DoHeatmap(tmp_so, 
               group.colors = discrete_palette_default,
               features = top_features,
               group.by = "clusters",
               angle = 0, 
               raster = FALSE, 
               draw.lines = TRUE) +
  scale_fill_gradientn(colours = viridis::viridis(256),
                       name = "Expression\nZ-scores")

save_plot(file.path(fig_dir, "heatmap_markers_myc_gfi1.pdf"), 
          p,
          base_width = 10,
          base_height = 12)

p
```

Assign cell_types with clustifyr

```{r}
library(clustifyr)
library(ComplexHeatmap)
tm_average <- readRDS("/Users/kriemo/Projects/sc_repos/vanotterloo/data/tabula_muris/TM_avg_expr.rds")

mdata <- get_metadata(so)

res <- clustify(so@assays$RNA@data, 
                tm_average, 
                query_genes = so@assays$RNA@var.features,
                metadata = mdata, 
                cluster_col = "clusters", 
                compute_method = "spearman")

hmap <- Heatmap(t(res), 
                viridis::viridis(256), 
                "Spearman\ncorrelation",
                row_title = "Cell types from Tabula Muris",
                column_title = "Clusters")

hmap

top_cell_types <- cor_to_call(res, threshold = 0.7)

so@meta.data <- so@meta.data %>% 
  tibble::rownames_to_column("cell") %>% 
  left_join(top_cell_types, 
            by = c("clusters" = "cluster")) %>% 
  dplyr::rename(tabula_muris_cell_type = type,
                cor_value = r) %>% 
  tibble::column_to_rownames("cell")

pdf(file.path(fig_dir, "heatmap_cell_type_cor_myc_gfi1.pdf"), 
          width = 10,
          height = 12)
draw(hmap)
dev.off()

```


### fix cell type assignments


```{r}
plot_umap(so, c("clusters", "tabula_muris_cell_type"), label_text = TRUE)
```


```{r}
new_ids <- c(
  '6' = 'Neuroendocrine',
  '4' = 'Neuroendocrine',
  '7' = 'Myleoid',
  '10' = 'Epithelial'
)

cids <- as.numeric(as.character(so$clusters)) %>% unique() %>% sort()

others <- setdiff(cids, names(new_ids))
other_ids <- rep("malignant", length(others))
names(other_ids) <- as.character(others)

new_ids <- c(new_ids, other_ids)

so$cell_types <- new_ids[as.character(so$clusters)]
```


```{r}
to_plot <- c("clusters", "cell_types", "tabula_muris_cell_type", "Phase") 
plts <- plot_umap(so, to_plot, label_text = TRUE, label_color = "black")

names(plts) <- to_plot
plts 

iwalk(plts, 
    ~save_plot(file.path(fig_dir, 
                         paste0("umap_gp34_myc_gfi1_by_", .y, ".pdf")),
          .x, 
          base_asp = 1.2))

```

## Markers just malignant


```{r}
so_sub <- subset(so, 
                 subset = cell_types %in% c("malignant", "Neuroendocrine"))

mkrs <- get_marker_summaries(so_sub, 
                             "clusters",
                             prefix = "myc_gfi1_malignant",
                     tsv_output_dir = mkrs_dir,
                     xlsx_output_dir = tbls_dir)
rm(mkrs)
```


## More detailed cell types

```{r}
relabel_clusters <- c(
  "0" = "MG-C1",
  "1" = "MG-C2",
  "2" = "MG-B1",
  "3" = "MG-C3",
  "4" = "MG-N1",
  "5" = "MG-B2",
  "6" = "MG-B3",
  "8" = "MG-N2",
  "9" = "MG-B4",
  "7" = "Immune",
  "10" = "Vasc.endothelium")


so$publication_cell_types <- relabel_clusters[so$clusters]


so$publication_cell_types <- factor(so$publication_cell_types,
                                    levels = relabel_clusters)


p <- plot_umap(so,
          "publication_cell_types", 
          label_text = TRUE,
          label_size = 4,
          legend_title = "",
          label_color = "black",
          ggrepel_opts = list(force = 0,
                              point.padding = 0.25))

p

save_plot(file.path(fig_dir, "umap_gfi1_cell_type_labels.pdf"),
          p,
          base_asp = 1.5)
```


```{r}
mkrs <- get_marker_summaries(so, 
                             "publication_cell_types",
                             prefix = "myc_gfi1_pub_cell_types",
                     tsv_output_dir = mkrs_dir,
                     xlsx_output_dir = tbls_dir)
mkrs


so_sub <- subset(so, 
                 subset = cell_types %in% c("malignant", "Neuroendocrine"))

mkrs <- get_marker_summaries(so_sub, 
                             "publication_cell_types",
                             prefix = "myc_gfi1_pub_cell_types_malignant",
                     tsv_output_dir = mkrs_dir,
                     xlsx_output_dir = tbls_dir)
rm(mkrs)
```

## Add in Northcutt paper metamodule scores

```{r}
if(!file.exists("../docs/northcutt_sup_tbl_2.xlsx")){
  supp_tbl_2 <- "https://static-content.springer.com/esm/art%3A10.1038%2Fs41586-019-1434-6/MediaObjects/41586_2019_1434_MOESM3_ESM.xlsx"

  download.file(supp_tbl_2, "../docs/northcutt_sup_tbl_2.xlsx")
}

sup_tbl <- read_excel("../docs/northcutt_sup_tbl_2.xlsx", 
           sheet = 2,
           skip = 3) 

sup_tbl[, 1] <- NULL

# extract out each metaprogram following their approach of excluding ribosomal
# and protein genes and non-coding RNAs

mps <- map(seq(1, ncol(sup_tbl), 2), 
    ~c(.x, .x + 1)) %>% 
  map(., ~sup_tbl[, .x] 
      %>% na.omit()) 

mp_names <- map_chr(mps, ~colnames(.x)[2])
mps <- map(mps, ~.x[, 2, drop = TRUE]) 
names(mps) <- mp_names

mps <- map(mps, 
    ~left_join(tibble(gene = .x),
               orthologs, 
               by = c("gene" = "external_gene_name")) %>% 
      pull(mmusculus_homolog_associated_gene_name) %>% 
      na.omit())

gp34_mps <- mps[str_subset(names(mps), "Group")]

mp_ids <- c(
  "Group 3/4-A_cell_cycle",
  "Group 3/4-B_Translation_Myc",
  "Group 3/4-C_Differentiation"
)

names(gp34_mps) <- mp_ids

for (i in seq_along(gp34_mps)){
  so <- AddModuleScore(so, 
                       features = list(c(gp34_mps[[i]])),
                       ctrl = 50,
                       name = names(gp34_mps)[i],
                       seed = 42)
}

new_ids <- str_c(make.names(mp_ids), "1") 

new_id_idx <- match(new_ids, colnames(so@meta.data))
colnames(so@meta.data)[new_id_idx] <- mp_ids

names(mp_ids) <- paste0("Hovestadt_et_al_", mp_ids)

plts <- imap(mp_ids,
            ~plot_feature(so, 
               .x, 
               embedding = "umap", 
               show_negative = TRUE, legend_title = "") + 
              labs(title = .y))

walk2(plts, 
      paste0("Hovestadt_et_al_", "Group 3 and 4-", c("A", "B", "C")),
    ~save_plot(file.path(fig_dir, 
                         paste0("umap_gp34_myc_gfi1_by_", .y, ".pdf")),
          .x, 
          base_asp = 1.4))

```

## Add our cluster modules

Using top marker genes found in our grp34 human clusters, assign scores to mouse model. 

```{r}
human_markers <- read_tsv(file.path("markers", "gp34",  "harmony_markers_gp3_only.tsv"))

# top 50 marker genes for modules and get mouse ortholog
human_modules <- filter(human_markers, padj < 0.05) %>% 
  group_by(group) %>% 
  arrange(padj, desc(logFC), .by_group = TRUE) %>% 
  left_join(orthologs, by = c("feature" = "external_gene_name")) %>% 
  na.omit() %>% 
  group_by(group) %>% 
  slice(1:50) %>% 
  split(., .$group) %>% 
  map(., ~pull(.x, mmusculus_homolog_associated_gene_name))

# dont't name anything with "luster" or "ouvain", as will not
# be recognized as float by ucsc cellbrowser...

names(human_modules) <- paste0("gp3_signatures_clust_",
                               unique(human_markers$group))

so <- AddModuleScore(so, 
                     features = human_modules,
                     ctrl = 50,
                     name = "tmp_",
                     seed = 42)

new_cols <- paste0("tmp_", 1:length(unique(human_markers$group)))
new_col_idx <- match(new_cols, colnames(so@meta.data))
better_cols <- names(human_modules)

# fix column names
colnames(so@meta.data)[new_col_idx] <- better_cols

names(better_cols) <- better_cols

plt <- imap(better_cols, 
    ~plot_umap(so, .x,
               show_negative = TRUE, 
               legend_title = "")) %>% 
  plot_grid(plotlist = ., 
            nrow = 3,
            ncol = 2)

plt 

save_plot(file.path(fig_dir, "umap_gp34_myc_gfi1_by_gp3_human_signatures.pdf"),
          plt, 
          base_asp = 1.4 ,
          nrow = 3,
          ncol = 2)
```



## Add our gene signature modules

Instead of naming as a cluster include subpopulation labels
```{r}
human_markers <- read_tsv(file.path("markers", "GP3",  "harmony_subpopulation_markers_GP3.tsv"))

# top 50 marker genes for modules and get mouse ortholog
human_modules <- filter(human_markers, padj < 0.05) %>% 
  group_by(group) %>% 
  arrange(padj, desc(logFC), .by_group = TRUE) %>% 
  left_join(orthologs, by = c("feature" = "external_gene_name")) %>% 
  na.omit() %>% 
  group_by(group) %>% 
  slice(1:200) %>% 
  split(., .$group) %>% 
  map(., ~pull(.x, mmusculus_homolog_associated_gene_name))

# dont't name anything with "luster" or "ouvain", as will not
# be recognized as float by ucsc cellbrowser...

names(human_modules) <- paste0("gp3_signatures_subpop_",
                               unique(human_markers$group))

so <- AddModuleScore(so, 
                     features = human_modules,
                     ctrl = 200,
                     name = "tmp_",
                     seed = 42)

new_cols <- paste0("tmp_", 1:length(unique(human_markers$group)))
new_col_idx <- match(new_cols, colnames(so@meta.data))
better_cols <- names(human_modules)

# fix column names
colnames(so@meta.data)[new_col_idx] <- better_cols

names(better_cols) <- better_cols

plt <- imap(better_cols, 
    ~plot_umap(so, .x,
               show_negative = TRUE, 
               legend_title = str_remove(.x, "gp3_signatures_subpop_"),
               cell_filter = colnames(so)[so$cell_types %in% c("malignant",
                                                               "Neuroendocrine")])) %>% 
  plot_grid(plotlist = ., 
            nrow = 3,
            ncol = 2)

plt 

save_plot(file.path(fig_dir, 
                    "umap_gfi1_by_gp3_human_subpopulation_signatures.pdf"),
          plt, 
          base_asp = 1.4 ,
          nrow = 3,
          ncol = 2)
```
```{r}
qsave(so, file.path(obj_dir,"gp3_myc_gfi1_sobj.qs"))
#so <- qread(file.path(obj_dir,"gp3_myc_gfi1_sobj.qs"))
```
## Calculate marker gene overlap

```{r}
library("GeneOverlap")

mouse_markers <- read_tsv(file.path("markers", "mouse", "gp34", "myc_gfi1_pub_cell_types_malignant_cluster_markers.tsv")) %>% 
filter(padj < 0.05) %>% 
  group_by(group) %>% 
  arrange(padj, desc(logFC), .by_group = TRUE) %>% 
  inner_join(orthologs, by = c("feature" = "mmusculus_homolog_associated_gene_name")) %>% 
    slice(1:200) %>% 
  split(., .$group) %>% 
  map(~pull(.x, external_gene_name))

human_markers <- read_tsv(file.path("markers", "GP3",  "harmony_subpopulation_markers_GP3.tsv")) %>% 
  filter(padj < 0.05) %>% 
  group_by(group) %>% 
  arrange(padj, desc(logFC), .by_group = TRUE) %>% 
  semi_join(orthologs, by = c("feature" = "external_gene_name")) %>% 
  slice(1:200) %>% 
  split(., .$group) %>% 
  map(~pull(.x, feature))


gom_human_mouse <- newGOM(mouse_markers, 
                       human_markers,
                       genome.size = nrow(so_sub))

# extract overlaps
pvals  <- getMatrix(gom_human_mouse, c("pval")) 
pvals <- apply(pvals, 2, function(x) p.adjust(x, n = length(pvals)))

jaccard <- getMatrix(gom_human_mouse, "Jaccard")
odds <- getMatrix(gom_human_mouse, c("odds.ratio"))

library(ComplexHeatmap)
# clip really low pvals to avoid take log10 of 0)
clip_val <- 1e-300
  
pvals[pvals < clip_val] <- clip_val
pvals <- -log10(pvals)

Heatmap(pvals,
        name = "-log10(pvalue)", 
        col = viridis::viridis(256),
        row_title = "Mouse model markers",
        column_title = "Human Group 3 and 4 markers")

h <- Heatmap(jaccard,
        name = "Jaccard index", 
        col = viridis::viridis(256),
       row_title = "Myc + GFI1 model clusters",
        column_title = "Human Group 3 and 4 clusters",
        cell_fun = function(j, i, x, y, width, height, fill) {
          bin <- findInterval(pvals[i, j], c(-log10(0.01), 20, 200))
          txt <- c("*", "**", "***")[bin]
          grid.text(txt, x, y, gp = gpar(fontsize = 10, col = "white"))
})


lgd_list <-  list(
    Legend(labels = c("*** < 1e-200",
                      "** < 1e-20",
                      "* < 1e-2"), 
           title = "Adj. p.val")
)
draw(h, heatmap_legend_list = lgd_list)


pdf(file.path(fig_dir, "Marker_overlap_gfi1_malignant_gp34.pdf"),
    width = 7,
    height = 7)
  draw(h, heatmap_legend_list = lgd_list)
dev.off()
```

### Make graph based on correlation 

```{r}
so_h <- qread(file.path("objects", "gp34", "gp3_only.qs"))
so_sub <- qread(file.path(obj_dir,"gp3_myc_gfi1_sobj.qs")) %>% 
  subset(., subset = cell_types %in% c("malignant", "Neuroendocrine"))

```

```{r}
mats <- set_shared_orthologs(so_h, so_sub, orthologs)
sos <- pmap(list(
  mats, 
  list(so_h, so_sub),
  c("tumor_subpopulations_gp3",
    "publication_cell_types")),
  function(mat, so, cluster_col) {
  tmp <- CreateSeuratObject(mat, meta.data = so@meta.data)
  Idents(tmp) <- cluster_col  
  avg_expr <- AverageExpression(tmp, return.seurat = FALSE)$RNA
  
  var_features <- VariableFeatures(FindVariableFeatures(tmp, nfeatures = 4000))
  
  list(avg_expr, var_features)
})
```

```{r}
shared_var_genes <- intersect(sos$mat1[[2]], sos$mat2[[2]])

h_var_mat <- sos$mat1[[1]][shared_var_genes, ] %>% 
  t() %>% 
  scale(scale = FALSE, center = TRUE) %>% 
  t()

m_var_mat <-  sos$mat2[[1]][shared_var_genes, ] %>% 
     t() %>% 
  scale(scale = FALSE, center = TRUE) %>% 
  t()

res <- list()
res$r <- matrix(nrow = ncol(m_var_mat),
                ncol = ncol(h_var_mat))
res$pvals <- res$r

for(m in seq_along(colnames(m_var_mat))){
  for(h in seq_along(colnames(h_var_mat))){
    s <- cor.test(m_var_mat[, m], h_var_mat[, h], method = "pearson")
    res$r[m, h] <- s$estimate
    res$pvals[m,h] <- s$p.value
  }
}

res$padj <- matrix(p.adjust(res$pvals),
                   nrow = nrow(res$pvals), 
                   ncol = ncol(res$pvals),
                   byrow = FALSE)
res <- map(res, ~{
  rownames(.x) <- colnames(m_var_mat)
  colnames(.x) <- colnames(h_var_mat)
  .x
})

```

```{r}
g <- graph.incidence(res$r, weighted = TRUE)
g <- delete_edges(g, which(E(g)$weight <= 0.2 | as.vector(t(res$padj)) >= 0.05))
wc <- cluster_walktrap(g)

V(g)$color <- V(g)$type
V(g)$shape <- ifelse(V(g)$type, "circle", "rectangle")
#plot(g, edge.width=E(g)$weight, layout=layout.fruchterman.reingold)

set.seed(2)
layout <-layout.fruchterman.reingold(g)

new_cols <- ifelse(startsWith(V(g)$name, "GP3"), "white", "black")

min_cor <- min(E(g)$weight)
max_cor <- max(E(g)$weight)
plot(wc, g,
     col = new_cols,
     mark.col = alpha(palette_OkabeIto[1:length(groups(wc))], 0.75),
     mark.border = "black",
     layout = layout,
    edge.width= scales::rescale(E(g)$weight, c(1,5)), 
     vertex.label = ifelse(startsWith(V(g)$name, "GP3"), 
                           str_remove(V(g)$name, "GP3-"), 
                           str_remove(V(g)$name, "MG-")),
     vertex.label.family = "sans",
    vertex.label.color = ifelse(startsWith(V(g)$name, "GP3"), "black", "white"))

legend("bottomleft",
       legend = c(paste("r =", formatC(min_cor, digits = 2)),
                  paste("r =", formatC(max_cor, digits = 2))),
       lwd = c(min(scales::rescale(E(g)$weight, c(1,5))),
               max(scales::rescale(E(g)$weight, c(1,5)))))
       

pdf(file.path(fig_dir, "gfi1_mouse_to_human_correlation_graph.pdf"))
min_cor <- min(E(g)$weight)
max_cor <- max(E(g)$weight)
plot(wc, g,
     col = new_cols,
     mark.col = alpha(palette_OkabeIto[1:length(groups(wc))], 0.75),
     mark.border = "black",
     layout = layout,
    edge.width= scales::rescale(E(g)$weight, c(1,5)), 
     vertex.label = ifelse(startsWith(V(g)$name, "GP3"), 
                           str_remove(V(g)$name, "GP3-"), 
                           str_remove(V(g)$name, "MG-")),
     vertex.label.family = "sans",
    vertex.label.color = ifelse(startsWith(V(g)$name, "GP3"), "black", "white"))

legend("bottomleft",
       legend = c(paste("r =", formatC(min_cor, digits = 2)),
                  paste("r =", formatC(max_cor, digits = 2))),
       lwd = c(min(scales::rescale(E(g)$weight, c(1,5))),
               max(scales::rescale(E(g)$weight, c(1,5)))))
       
dev.off()
```

## Write out metadata


```{r} 
mdata_out <- get_metadata(so)
outname <- file.path(tbls_dir, 
                     paste0("gp3_myc_gfi1_metadata_", 
                            format(Sys.Date(), "%Y_%m_%d"), 
                            ".tsv.gz"))
write_tsv(mdata_out, outname)
```

## save object

```{r}
#qsave(so, file.path(obj_dir,"gp3_myc_gfi1_sobj.qs"))
#so <- qread(file.path(obj_dir,"gp3_myc_gfi1_sobj.qs"))
```

# Integrate with human


```{r}
so <- qread(file.path(obj_dir,"gp3_myc_gfi1_sobj.qs"))

int_dir <- "human_integrated"
fig_dir <- file.path(fig_dir, int_dir)
mkrs_dir <- file.path(mkrs_dir, int_dir)
tbls_dir <- file.path(tbls_dir, int_dir)
obj_dir <- file.path(obj_dir, int_dir)

walk(c(fig_dir, mkrs_dir, tbls_dir, obj_dir),
     dir.create, showWarnings = F)
```

```{r load_human_dat}
so_m <- subset(so, subset = cell_types %in% c("malignant", "Neuroendocrine"))
so_h <- qread(file.path("objects", "gp34", "gp34.qs"))
```

```{r covert mouse to human orths}
# use shared genes based on ortholog table
count_mat <- so_m@assays$RNA@counts
norm_mat <- so_m@assays$RNA@data

new_gids <- left_join(tibble(genes = rownames(count_mat)),
                      orthologs, 
                      by = c("genes" = "mmusculus_homolog_associated_gene_name"))
  

genes_to_keep <- which(!is.na(new_gids$external_gene_name))

count_mat <- count_mat[genes_to_keep, ]
norm_mat <- norm_mat[genes_to_keep, ]

rownames(count_mat) <- new_gids$external_gene_name[genes_to_keep]
rownames(norm_mat) <- new_gids$external_gene_name[genes_to_keep]

so_m <- CreateSeuratObject(count_mat, 
                   meta.data = so_m@meta.data)

so_m <- SetAssayData(so_m, "data", new.data = norm_mat)

shared_ids <- intersect(rownames(so_m), rownames(so_h))

so_m <- so_m[shared_ids, ]
so_h <- so_h[shared_ids, ]

so_m$subgroup <- ifelse(str_detect(so_m$UPN, "^MED"),
                        "shh_model",
                        ifelse(so_m$UPN == "2_MP",
                               "MYC_DNP53_model",
                               ifelse(
                               so_m$UPN == "3_MG",
                               "MYC_GFI1_model",
                               NA)))
  
so_m$species <- "mouse"
so_h$species <- "human"

so <- merge(so_h, so_m)

rm(so_h, so_m); gc()
```


```{r}
res_settings <- c(0.1, 0.2, 0.3, 0.4, 0.5, 0.6)
so <- FindVariableFeatures(so, 
                       selection.method = "vst", 
                       nfeatures = 3000, 
                       verbose = FALSE) %>% 
  ScaleData(verbose = FALSE) %>% 
  RunPCA(npcs = 40, 
         verbose = FALSE, 
         seed.use = seed_value) %>% 
  FindNeighbors(reduction = "pca", 
                dims = 1:30,
                verbose = FALSE) %>% 
  RunUMAP(dims = 1:30,
          min.dist = 0.3,
          seed.use = seed_value,
          verbose = FALSE) %>% 
  FindClusters(resolution = res_settings,
               verbose = FALSE) 

so$seurat_clusters <- NULL
```



```{r}
to_plot <- c("species", "subgroup")
plts <- plot_umap(so, to_plot) 
names(plts) <- to_plot

iwalk(plts, 
    ~save_plot(file.path(fig_dir, 
                         paste0("umap_gp34_gfi1_by_", .y, ".pdf")),
          .x, 
          base_asp = 1.5))

plt <- plot_grid(plotlist = plts,
                 nrow = 1, 
                 ncol = 2)

save_plot(file.path(fig_dir, 
                    paste0("umap_gp34_gfi1_summary.pdf")),
          plt,
          nrow = 1, 
          ncol = 2, 
          base_asp = 1.5)
plt
```

## Harmonize by sample

```{r}
library(harmony)
so <- RunHarmony(so, group.by.vars = "UPN", 
                 reduction.save = "harmony_upn")

so <- RunUMAP(so,
              reduction = "harmony_upn",
              dims = 1:30,
              reduction.name = "harmony_umap_upn")

so <- FindNeighbors(so,
                    reduction = "harmony_upn",
                    dims = 1:20)

to_plot <- c("species", "subgroup", "coarse_clusters_gp34_harmony")
plts <- plot_feature(so, 
                     to_plot, 
                     sorted = "random", 
                     embedding = "harmony_umap_upn") 
names(plts) <- to_plot

iwalk(plts, 
    ~save_plot(file.path(fig_dir, 
                         paste0("harmony_upn_umap_gp34_gfi1_by_", .y, ".pdf")),
          .x, 
          base_asp = 1.5))

plt <- plot_grid(plotlist = plts,
                 nrow = 1, 
                 ncol = 2)

save_plot(file.path(fig_dir, 
                    paste0("harmony_upn_umap_gp34_gfi1_summary.pdf")),
          plt,
          nrow = 1, 
          ncol = 2, 
          base_asp = 1.5)
plt
```

### Human signature overlap

#### gfi1 model
```{r}
Idents(so) <- "UPN"
mouse_model_cells <- WhichCells(so, 
                                idents = "3_MG")
other_cells <- WhichCells(so,
                          idents = "3_MG", 
                          invert = TRUE)

library(RANN)
kval <- 10
ndims <- 1:20
nn_vals <- nn2(
  so@reductions$harmony_upn@cell.embeddings[other_cells, ndims],
  so@reductions$harmony_upn@cell.embeddings[mouse_model_cells, ndims],
  k = kval)
    
nn_cells <- tibble(
  mouse_cell = rep(mouse_model_cells, each = kval),
  n_idx = rep(1:kval, length(mouse_model_cells)),
  human_neighbor = other_cells[nn_vals$nn.idx])

nn_cells <- get_metadata(so, embedding = NULL) %>%
  select(cell = cell, coarse_clusters_gp34_harmony) %>% 
  filter(cell %in% other_cells) %>% 
  left_join(nn_cells, ., by = c("human_neighbor" = "cell"))

nn_cell_summary <- nn_cells %>%
  mutate(coarse_clusters_gp34_harmony = factor(coarse_clusters_gp34_harmony)) %>%
  group_by(mouse_cell, coarse_clusters_gp34_harmony, .drop = FALSE) %>% 
  tally() %>% 
  group_by(mouse_cell) %>% 
  mutate(cluster_rank = rank(n, ties.method = "random")) %>%
  arrange(desc(cluster_rank), .by_group = TRUE) %>% 
  slice(1) %>% 
  ungroup()

mouse_gfi1_counts <- nn_cell_summary %>% 
  group_by(coarse_clusters_gp34_harmony, 
                             .drop = FALSE) %>%
  tally() %>% 
  mutate(id = "Myc_Gfi1_mouse_model") %>% 
  pivot_wider(names_from = "coarse_clusters_gp34_harmony",
              values_from = "n") %>% 
  column_to_rownames("id")
```

#### dnp53
```{r}

so_p53 <- qread(file.path(obj_dir, "human_gp34_myc_dnp53_sobj.qs"))

Idents(so_p53) <- "UPN"
mouse_model_cells <- WhichCells(so_p53, 
                                idents = "2_MP")
other_cells <- WhichCells(so_p53,
                          idents = "2_MP", 
                          invert = TRUE)

kval <- 10
ndims <- 1:20
nn_vals <- nn2(
  so_p53@reductions$harmony_upn@cell.embeddings[other_cells, ndims],
  so_p53@reductions$harmony_upn@cell.embeddings[mouse_model_cells, ndims],
  k = kval)
    
nn_cells <- tibble(
  mouse_cell = rep(mouse_model_cells, each = kval),
  n_idx = rep(1:kval, length(mouse_model_cells)),
  human_neighbor = other_cells[nn_vals$nn.idx])

nn_cells <- get_metadata(so, embedding = NULL) %>%
  select(cell = cell, coarse_clusters_gp34_harmony) %>% 
  filter(cell %in% other_cells) %>% 
  left_join(nn_cells, ., by = c("human_neighbor" = "cell"))

nn_cell_summary <- nn_cells %>%
  mutate(coarse_clusters_gp34_harmony = factor(coarse_clusters_gp34_harmony)) %>%
  group_by(mouse_cell, coarse_clusters_gp34_harmony, .drop = FALSE) %>% 
  tally() %>% 
  group_by(mouse_cell) %>% 
  mutate(cluster_rank = rank(n, ties.method = "random")) %>%
  arrange(desc(cluster_rank), .by_group = TRUE) %>% 
  slice(1) %>% 
  ungroup()

mouse_dnp53_counts <- nn_cell_summary %>% 
  group_by(coarse_clusters_gp34_harmony, 
                             .drop = FALSE) %>%
  tally() %>% 
  mutate(id = "Myc_dnp53_mouse_model") %>% 
  pivot_wider(names_from = "coarse_clusters_gp34_harmony",
              values_from = "n") %>% 
  column_to_rownames("id")
```

```{r}
human_counts <- get_cell_count_matrix(so_h, 
                      "UPN", 
                      "coarse_clusters_gp34_harmony")

# use hclustering to order samples
sample_order <- rownames(human_counts)[hclust(dist(human_counts / rowSums(human_counts)))$order]
sample_order <- c(sample_order, "Myc_Gfi1_mouse_model", "Myc_dnp53_mouse_model")

combined_counts <- rbind(human_counts,
                             mouse_gfi1_counts) %>% 
  rbind(., mouse_dnp53_counts) %>% 
  rownames_to_column("sample_id") %>% 
  pivot_longer(cols = -sample_id)

plt_dat <- group_by(combined_counts, sample_id) %>% 
      mutate(n_cells = sum(value)) %>%
      group_by(sample_id, name) %>%
      summarize(prop_cell_type = value / unique(n_cells)) %>% 
  ungroup()

sup_data <- get_metadata(so_h) %>% 
  select(UPN, subgroup) %>% 
  unique() 

plt_dat <- left_join(plt_dat, sup_data, by = c("sample_id" =  "UPN")) %>% 
  mutate(subgroup = ifelse(is.na(subgroup), 
                           "mouse",
                           subgroup)) %>% 
  mutate(sample_id = factor(sample_id, levels = sample_order))
  
  
cell_summary <- group_by(combined_counts, sample_id) %>% 
    mutate(n_cells = sum(value)) %>% 
  ungroup() %>%
  select(sample_id, name, n_cells) %>% 
  mutate(n_cells = str_c("n = ", scales::comma(n_cells)),                              n_cells = str_pad(n_cells, max(nchar(n_cells)), "right")) %>%
    unique() %>% 
   mutate(sample_id = factor(sample_id, levels = sample_order))

color_map <- select(plt_dat, sample_id, subgroup) %>% 
  unique() %>% 
  pull(subgroup)
names(color_map) <- select(plt_dat, sample_id, subgroup) %>% 
  unique() %>% 
  pull(sample_id)

color_map <- color_map[sample_order]
color_map <- palette_OkabeIto[as.integer(factor(color_map))]
p <- ggplot(plt_dat,
            aes(sample_id, prop_cell_type)) +
    geom_col(aes(fill = name)) +
    labs(x = "Sample ID",
         y = "Proportion of each cell type")  + 
  scale_fill_manual(values = discrete_palette_default) + 
  labs(fill = "Human Gp34\nclusters",
       x = "") + 
  theme(axis.text.x = element_text(angle = 90, 
                                   hjust = 1, 
                                   vjust = 0.5),
          legend.position = "top",
          strip.background = element_rect(fill = "white")) +
  geom_text(data = cell_summary,
            aes(x = sample_id, y = 0.15,
                  label = n_cells),
            angle = 90)

p

save_plot(file.path(fig_dir, "cell_proportions_in_human_clusters.pdf"), 
          p,
          base_asp = 1.1,
          base_height = 6)
```

## Harmonize by species

```{r}
library(harmony)

so <- RunHarmony(so, group.by.vars = "species")
so <- RunUMAP(so,
              reduction = "harmony",
              dims = 1:30,
              reduction.name = "harmony_umap")
so <- FindNeighbors(so,
                    reduction = "harmony",
                    dims = 1:20)

to_plot <- c("species", "subgroup", "coarse_clusters_gp34_harmony")
plts <- plot_harmony(so, to_plot) 
names(plts) <- to_plot

iwalk(plts, 
    ~save_plot(file.path(fig_dir, 
                         paste0("harmony_species_umap_gp34_gfi1_by_", .y, ".pdf")),
          .x, 
          base_asp = 1.5))

plt <- plot_grid(plotlist = plts,
                 nrow = 1, 
                 ncol = 2)

save_plot(file.path(fig_dir, 
                    paste0("harmony_species_umap_gp34_gfi1_summary.pdf")),
          plt,
          nrow = 1, 
          ncol = 2, 
          base_asp = 1.5)
plt
```

## save object

```{r}
qsave(so, file.path(obj_dir,"human_gp34_myc_gfi1_sobj.qs"))
#so <- qread(file.path(obj_dir, "human_gp34_myc_gfi1_sobj.qs"))
```


# Extract out immune / normal cells

```{r}
source(here::here("/R/utils.R"))
int_dir <- rmd_dir <- file.path("mouse", "gp34", "normal_immune")
fig_dir <- file.path(fig_dir, int_dir)
mkrs_dir <- file.path(mkrs_dir, int_dir)
tbls_dir <- file.path(tbls_dir, int_dir)
obj_dir <- file.path(obj_dir, int_dir)

walk(c(fig_dir, mkrs_dir, tbls_dir, obj_dir),
     dir.create, showWarnings = F)
```

```{r}
so <- qread("objects/mouse/gp34/gp3_myc_gfi1_sobj.qs")

so <- subset(so, subset = cell_types %in% c("Myleoid"))

res_settings <- seq(0.1, 0.7, 0.1)
so <- FindVariableFeatures(so, nfeatures = 2000) %>% 
  ScaleData(verbose = FALSE) %>% 
  RunPCA(npcs = 40, 
         verbose = FALSE, 
         seed.use = seed_value) %>% 
  FindNeighbors(reduction = "pca", 
                dims = 1:20,
                verbose = FALSE) %>% 
  RunUMAP(dims = 1:20,
          min.dist = 0.3,
          seed.use = seed_value,
          verbose = FALSE) %>% 
  FindClusters(resolution = res_settings,
               random.seed = seed_value,
               verbose = FALSE)
```

```{r}
cluster_cols <- str_subset(colnames(so@meta.data), "RNA_snn_res")

plts <- plot_umap(so, cluster_cols)
names(plts) <- cluster_cols

iwalk(plts, 
    ~save_plot(file.path(fig_dir, 
                         paste0("umap_gp34_gfi1_by_", .y, ".pdf")),
          .x, 
          base_asp = 1))

plt <- plot_grid(plotlist = plts,
                 nrow = 3, 
                 ncol = 3)

save_plot(file.path(fig_dir, 
                         paste0("umap_gp34_gfi1_all_clusters.pdf")),
          plt,
          nrow = 3, 
          ncol = 3, 
          base_asp = 1.5)
plt
```

```{r}
so$clusters <- so$RNA_snn_res.0.6

library(presto)

mkrs <- get_marker_summaries(so, 
                             "clusters",
                             prefix = "myc_gfi1",
                             tsv_output_dir = mkrs_dir,
                             xlsx_output_dir = tbls_dir)
mkrs
```

Assign cell_types with clustifyr

```{r}
library(clustifyr)
library(ComplexHeatmap)
tm_average <- readRDS("/Users/kriemo/Projects/sc_repos/vanotterloo/data/tabula_muris/TM_avg_expr.rds")

mdata <- get_metadata(so)

res <- clustify(so@assays$RNA@data, 
                tm_average, 
                query_genes = so@assays$RNA@var.features,
                metadata = mdata, 
                cluster_col = "clusters", 
                compute_method = "spearman")

hmap <- Heatmap(t(res), 
                viridis::viridis(256), 
                "Spearman\ncorrelation",
                row_title = "Cell types from Tabula Muris",
                column_title = "Clusters")

hmap

top_cell_types <- cor_to_call(res, threshold = 0.5)

so@meta.data <- so@meta.data %>% 
  tibble::rownames_to_column("cell") %>% 
  left_join(top_cell_types, 
            by = c("clusters" = "cluster")) %>% 
  dplyr::mutate(tabula_muris_cell_type = type,
                cor_value = r) %>% 
  dplyr::select(-type, -r) %>% 
  tibble::column_to_rownames("cell")

pdf(file.path(fig_dir, "heatmap_cell_type_cor_myc_gfi1.pdf"), 
          width = 10,
          height = 12)
draw(hmap)
dev.off()



p <- plot_umap(so, "tabula_muris_cell_type")
save_plot(file.path(fig_dir,"umap_gp34_gfi1_by_tm_cell_type.pdf"),
          p, 
          base_asp = 1.5)
p
```

Tweak cell type assignments

```{r}
new_ids <- c(
  "0" = "other myeloid"
)

so$cell_types <- ifelse(as.character(so$clusters) %in% names(new_ids),
                        new_ids[as.character(so$clusters)],
                        so$tabula_muris_cell_type)

plot_umap(so, "cell_types")
```

```{r}
mkrs <- get_marker_summaries(so, 
                             "cell_types",
                             prefix = "myc_gfi1_celltype",
                             tsv_output_dir = mkrs_dir,
                             xlsx_output_dir = tbls_dir)
mkrs
```

## save object

```{r}
qsave(so, file.path(obj_dir,"gp3_myc_gfi1_sobj.qs"))
```
