---
title: "Format UCSC cellbrowser"
date: "`r Sys.Date()`"
knit: (function(inputFile, encoding) {
  rmarkdown::render(inputFile, encoding = encoding, output_dir = "html") })
output:
  html_document:
    toc: true
    toc_float: true
    toc_collapsed: false
    theme: cosmo
    highlight: tango
    fig_caption: true
    code_folding: hide
    df_print: paged
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```


Format seurats object into a UCSC cellbrowser. Note requires UCSC (`cellbrowser python package`[ https://cellbrowser.readthedocs.io/index.html]) 

```{r}
source(here::here("R/utils.R"))
library(scbp)
library(tidyverse)
library(qs)
library(Seurat)
library(glue)
library(SeuratWrappers)

cb_outdir <- file.path("cellbrowser", "publication-final")
walk(file.path(cb_outdir, c("mouse", "human")), ~dir.create(.x, recursive = TRUE, showWarnings = FALSE))

alra_objs <- file.path("objects", "alra")
dir.create(alra_objs, recursive = TRUE, showWarnings = FALSE)

def_embeddings <- function(so) {str_subset(names(so@reductions), "umap")}

write_config <- function(x, path){
  glue('{key}="{value}"', 
     key = names(x), 
     value = x) %>% 
    write_lines(path)
}

qread <- function(file, ...) {qs::qread(file, ..., nthreads = 4, use_alt_rep = FALSE)}
```

# Collection overview

```{r}
collection_cb <- c(shortLabel="single cell survey of mouse and human medulloblastoma")

collection_desc <- c(title="Single cell survey of childhood Medulloblastoma",
                     abstract="We explored cellular heterogeneity in medulloblastoma (MB) using single-cell RNA sequencing (scRNA-seq), immunohistochemistry and deconvolution of bulk transcriptomic data. Over 45,000 cells from 28 patients from all main subgroups of medulloblastoma (1 WNT, 9 SHH, 7 GP3 and 11 GP4) were clustered using Harmony alignment to identify conserved subpopulations. Each subgroup contained subpopulations exhibiting mitotic, undifferentiated and neuronal differentiated transcript profiles, corroborating other recent medulloblastoma scRNA-seq studies. The magnitude of our present study builds on the findings of existing studies, and characterizes new  conserved neoplastic subpopulations. We uncovered a photoreceptor-differentiated subpopulation that was predominantly found in GP3 medulloblastoma, and in SHH, a subpopulation that constitutes differentiating nodules. Deconvolution of a large transcriptomic dataset showed that neoplastic subpopulations are associated with major and minor subgroup subdivisions, for example, photoreceptor subpopulation cells are more abundant in GP3-alpha. This scRNA-seq dataset also afforded unique insights into the diverse immune landscape of medulloblastoma, and revealed an SHH nodule-associated myeloid subpopulation. Additionally, we performed scRNA-seq on 16,000 cells from genetically engineered mouse (GEM) models of GP3 and SHH medulloblastoma. These models showed a level of fidelity with corresponding human subgroup-specific neoplastic subpopulations. Collectively, our findings advance our understanding of the neoplastic and immune landscape of the main medulloblastoma subgroups in both humans and GEM models.",
                     unitDesc="Log-normalized counts",
                     geo_series="GSE155446",
                     image="MAF_Logo_Purple_Tagline-300x163.png",
                     doi="https://doi.org/10.1093/neuonc/noab135",
                     other_url="https://github.com/rnabioco/medulloblast")

write_config(collection_cb, file.path(cb_outdir, "cellbrowser.conf"))
write_config(collection_desc, file.path(cb_outdir, "desc.conf"))
```


# Human tumors

```{r}
collection_cb <- c(shortLabel="Childhood MB (WNT,GP3,GP4,SHH)")

collection_desc <- c(title="Childhood Medulloblastoma",
                     abstract="single cell RNA-seq of childhood MB (1 WNT, 9 SHH, 7 GP3 and 11 GP4). See individual datasets for a description of dataset content")

write_config(collection_cb, file.path(cb_outdir, "human", "cellbrowser.conf"))
write_config(collection_desc, file.path(cb_outdir, "human", "desc.conf"))
```

## All cells


```{r}
so <- qread(file.path("objects", "preprocess", "so.qs"))

so$subgroup <- ifelse(so$subgroup == "GP3/4", "GP3", so$subgroup)

cols_to_keep <- c(
  `genes per cell` = "nFeature_RNA",
  `UMIs per cell` = "nCount_RNA",
  `percent of UMIs as mitochondria` = "percent.mt",
  `clusters` = "coarse_clusters",
  `subgroup` = "subgroup",
  `Dx` = "Dx",
  `UPN` = "UPN",
  `other_sample_info` = "additional_info",
   cell_type = "coarse_cell_type",
  `cluster_diversity` = "cdiversity")

markdown::markdownToHTML(text = c(
                   "Full dataset, includes all cells.",
                   "",
                   "##### Annotation descriptions:  ",
                   "",
                   "* subgroup: Tumor subgroup",
                   "* UPN: Unique patient identifier",
                   "* Other sample info: Information on primary or recurrence status",
                   "* cell type: Inferred coarsely defined cell type ",
                   "* cluster diversity: Entropy measure of diversity of patient samples in each cluster.  0 = cluster derived from 1 patient, 1 = cluster is an equal mix from all patients "
                   
                   ),
  output = file.path(cb_outdir, "human", "All-cells", "abstract.html"),
  fragment.only = TRUE)
 
make_cellbrowser(so, 
                 column_list = cols_to_keep,
                 secondary_cols = c(
                    "subgroup",
                    "cell_type"), 
                 project = "All-cells",
                 outdir = file.path(cb_outdir, "human"),
                 marker_file = "markers/preprocess/coarse_cluster_markers_all_data.tsv",
                 ident = "clusters",
                 embeddings = def_embeddings(so),
                 skip_expr_matrix = TRUE,
                 config = list(priority = 1,  
                               radius = 2),
                 description = list(    
                   title = "Childhood Medulloblastoma",
                   abstractFile = "abstract.html"
                 )
)
```



## SHH subgroup

```{r}
so <- qread(file.path("objects", "shh", "shh.qs"))

so$tumor_cell_type_shh <- str_replace_all(so$tumor_cell_type_shh, ",", "")
so$tumor_cell_types_all_subgroups <-  str_replace_all(so$tumor_cell_types_all_subgroups, ",", " ")

cols_to_keep <- c(
  `genes per cell` = "nFeature_RNA",
  `UMIs per cell` = "nCount_RNA",
  `percent of UMIs as mitochondria` = "percent.mt",
  `clusters_unaligned` = "coarse_clusters_shh",
  `clusters_aligned` = "coarse_clusters_shh_harmony",
  `subtype` = "subtype",
  `UPN` = "UPN",
  `other_sample_info` = "additional_info",
   cell_type = "coarse_cell_type",
   `cell cycle phase` = "Phase",
  "Northcutt_SHH-A-cell_cycle" = "SHH-A-cell_cycle",
  "Northcutt_SHH-B-SHH_signaling" = "SHH-B-SHH_signaling",
  "Northcutt_SHH-C-Differentiation" = "SHH-C-Differentiation",
  "CytoTRACE-GCS" = "GCS",
  "CytoTRACE-score" = "CytoTRACE",
  unaligned_cluster_diversity = "unaligned_cluster_diversity", 
  aligned_cluster_diversity = "aligned_cluster_diversity",
  `Subpopulations` = "tumor_subpopulations_shh",
  `CNV count (methylation)` = "methyl_nCNV"
  )

so@assays$alra <- get_alra_assay(so,
                                 file.path(alra_objs, "shh.qs"),
                                 overwrite = FALSE)
dir.create(file.path(cb_outdir, "human/SHH-neoplastic-cells"), showWarnings = FALSE, recursive = TRUE)
markdown::markdownToHTML(text = c(
                   "Neoplastic cells from SHH subgroup.  ",
                   "There are both unaligned (umap) and aligned (harmony_umap) projections  ",
                   "",
                   "##### Annotation description:  ",
                   "",
                   "* clusters unaligned: clusters derived from PCA + Louvain clustering",
                   "* clusters aligned: clusters derived from Harmony + Louvain clustering",
                   "* subtype: The highest correlated MB subtype based on classifications from Cavalli FMG, Remke M, Rampasek L, et al. Intertumoral Heterogeneity within Medulloblastoma Subgroups. Cancer Cell. 2017;31:737–754.e6",
                   "* UPN: Unique patient identifier",
                   "* cell cycle phase: inferred cell cycle phase",
                   "* Northcutt SHH-*: Gene modules generated from subpopulations described in Hovestadt V, Smith KS, Bihannic L, et al. Resolving medulloblastoma cellular architecture by single-cell genomics. Nature. 2019;572:74–79.",
                   "* Cytotrace-*: Scores derived by applying the cytotrace algorithm",
                   "* cluster diversity: Entropy measure of tumor diversity within each cluster, 0 = cluster derived from 1 patient, 1 = cluster equal mix from all patients.",
                   "* Subpopulations: subpopulations described in manuscript, these annotations are displayed in the browser, and the markers displayed match the harmony_umap layout",
                   "* CNV count (methylation): # of CNVs based on methylation data  ",
                   "",
                   "##### Feature descriptions",
                   "",
                   "* CNV calls per cell based on inferCNV have been added to the data matrix, which can be viewed by querying the 'Color by gene' tab for cnv metrics. The metrics are prefixed with 'CNV_' (e.g. CNV_has-cnv-chr9, CNV_proportion_cnv-chr22)  ",
                   "  +  The fields has_cnv/has_loss/has_dupli are 0/1 boolean answers to whether respectively any kind of cnv, a loss or a duplication is found within the given chr.  ",
"  +  proportion_cnv/proportion_loss/proportion_dupli is the proportion in number of genes that are part of any cnv/loss cnv/duplication cnv in the given chr.  ",
"  +  proportion_scaled are similar to proportion but taking into account whether genes that are duplications/losses are of a single copy or two copies as a weight  ",
"  +  top_loss_n/top_dupli_n are the n loss/duplication cnvs that cover the most genes  ",
                   "* ALRA imputed values are also present for visualization purposes and can be queried using 'ALRA_gene_name' (e.g. ALRA_GRIA2) "),
  output = file.path(cb_outdir, "human", "SHH-neoplastic-cells", "abstract.html"),
  fragment.only = TRUE)

make_cellbrowser(so, 
                 column_list = cols_to_keep,
                 secondary_cols = 
                   c("subtype",
                     "cell_type",
                     "subgroup",
                     "cell cycle phase"), 
                 project = "SHH-neoplastic-cells",
                 outdir = file.path(cb_outdir, "human"),
                 marker_file = "markers/shh/harmony_subpopulation_markers_SHH.tsv",
                 ident = "Subpopulations",
                 embeddings = def_embeddings(so),
                 skip_expr_matrix = TRUE, 
                 assays = c("ALRA_" = "alra",
                            "CNV_" = "CNV"),
                  config = list(priority = 3,  
                               radius = 2),
                 description = list(    
                   title = "Medulloblastoma SHH subgroup",
                   abstractFile = "abstract.html" 
                 )
)
```



## Gp3 subgroup


```{r}
so <- qread(file.path("objects", "gp34", "gp3_only.qs"))
so$subgroup <- ifelse(so$subgroup == "GP3/4", "GP3", so$subgroup)
so$tumor_cell_type_gp3 <- str_replace_all(so$tumor_cell_type_gp3, ",", "")
so$tumor_cell_types_all_subgroups <-  str_replace_all(so$tumor_cell_types_all_subgroups, ",", " ")

cols_to_keep <- c(
  `genes per cell` = "nFeature_RNA",
  `UMIs per cell` = "nCount_RNA",
  `percent of UMIs as mitochondria` = "percent.mt",
  `clusters_unaligned` = "coarse_clusters_gp3",
  `clusters_aligned` = "coarse_clusters_gp3_harmony",
  `subgroup` = "subgroup",
  `subtype` = "subtype",
  `Dx` = "Dx",
  `UPN` = "UPN",
    `other_sample_info` = "additional_info",
   `cell cycle phase` = "Phase",
   "Northcutt_Group 3/4-A-cell_cycle" = "Group 3/4-A-cell_cycle",
   "Northcutt_Group 3/4-B-Translation_Myc" = "Group 3/4-B-Translation_Myc",
   "Northcutt_Group 3/4-C-Differentiation" = "Group 3/4-C-Differentiation",
  "CytoTRACE-GCS" = "GCS",
  "CytoTRACE-score" = "CytoTRACE",
  unaligned_cluster_diversity = "unaligned_cluster_diversity", 
  aligned_cluster_diversity = "aligned_cluster_diversity",
  `Subpopulations` = "tumor_subpopulations_gp3",
  `CNV count (methylation)` = "methyl_nCNV"
  )


so@assays$alra <- get_alra_assay(so,
                                 file.path(alra_objs, "gp3.qs"),
                                 overwrite = FALSE)

dir.create(file.path(cb_outdir, "human", "GP3-neoplastic-cells"),
           showWarnings = FALSE,
           recursive = TRUE)
           
markdown::markdownToHTML(text = c(
                   "Neoplastic cells from Group 3 subgroup.  ",
                   "There are both unaligned (umap) and aligned (harmony_umap) projections  ",
                   "",
                   "##### Annotation description:  ",
                   "",
                   "* clusters unaligned: clusters derived from PCA + Louvain clustering",
                   "* clusters aligned: clusters derived from Harmony + Louvain clustering",
                   "* subtype: The highest correlated MB subtype based on classifications from Cavalli FMG, Remke M, Rampasek L, et al. Intertumoral Heterogeneity within Medulloblastoma Subgroups. Cancer Cell. 2017;31:737–754.e6",
                   "* UPN: Unique patient identifier",
                   "* cell cycle phase: inferred cell cycle phase",
                   "* Northcutt SHH-*: Gene modules generated from subpopulations described in Hovestadt V, Smith KS, Bihannic L, et al. Resolving medulloblastoma cellular architecture by single-cell genomics. Nature. 2019;572:74–79.",
                   "* Cytotrace-*: Scores derived by applying the cytotrace algorithm",
                   "* cluster diversity: Entropy measure of tumor diversity within each cluster, 0 = cluster derived from 1 patient, 1 = cluster equal mix from all patients.",
                   "* Subpopulations: subpopulations described in manuscript, these annotations are displayed in the browser, and the markers displayed match the harmony_umap layout",
                   "* CNV count (methylation): # of CNVs based on methylation data  ",
                   "",
                   "##### Feature descriptions",
                   "",
                   "* CNV calls per cell based on inferCNV have been added to the data matrix, which can be viewed by querying the 'Color by gene' tab for cnv metrics. The metrics are prefixed with 'CNV_' (e.g. CNV_has-cnv-chr9, CNV_proportion_cnv-chr22)  ",
                   "  +  The fields has_cnv/has_loss/has_dupli are 0/1 boolean answers to whether respectively any kind of cnv, a loss or a duplication is found within the given chr.  ",
"  +  proportion_cnv/proportion_loss/proportion_dupli is the proportion in number of genes that are part of any cnv/loss cnv/duplication cnv in the given chr.  ",
"  +  proportion_scaled are similar to proportion but taking into account whether genes that are duplications/losses are of a single copy or two copies as a weight  ",
"  +  top_loss_n/top_dupli_n are the n loss/duplication cnvs that cover the most genes  ",
                   "* ALRA imputed values are also present for visualization purposes and can be queried using 'ALRA_gene_name' (e.g. ALRA_GRIA2) "),
  output = file.path(cb_outdir, "human", "GP3-neoplastic-cells", "abstract.html"),
  fragment.only = TRUE)


scbp::make_cellbrowser(so, 
                 column_list = cols_to_keep,
                 secondary_cols = 
                   c("subtype",
                     "cell_type",
                     "subgroup",
                     "cell cycle phase"), 
                 project = "GP3-neoplastic-cells",
                 outdir = file.path(cb_outdir, "human"),
                 marker_file = "markers/gp3/harmony_subpopulation_markers_gp3.tsv",
                 ident = "Subpopulations",
                 embeddings = def_embeddings(so),
                 skip_expr_matrix = TRUE,
                 assays = c("ALRA_" = "alra",
                            "CNV_" = "CNV"),
                   config = list(priority = 5),
                 description = list(    
                   title = "Medulloblastoma Group 3 subgroup",
                   abstractFile = "abstract.html"
                 )
)
```



## Gp4 subgroup

```{r}
so <- qread(file.path("objects", "gp34", "gp4_only.qs"))

so$tumor_cell_type_gp4 <- str_replace_all(so$tumor_cell_type_gp4, ",", "")
so$tumor_cell_types_all_subgroups <-  str_replace_all(so$tumor_cell_types_all_subgroups, ",", " ")

cols_to_keep <- c(
  `genes per cell` = "nFeature_RNA",
  `UMIs per cell` = "nCount_RNA",
  `percent of UMIs as mitochondria` = "percent.mt",
  `clusters_unaligned` = "coarse_clusters_gp4",
  `clusters_aligned` = "coarse_clusters_gp4_harmony",
  `subgroup` = "subgroup",
  `subtype` = "subtype",
  `Dx` = "Dx",
  `other_sample_info` = "additional_info",
  `UPN` = "UPN",
   `cell cycle phase` = "Phase",
   "Northcutt_Group 3/4-A-cell_cycle" = "Group 3/4-A-cell_cycle",
   "Northcutt_Group 3/4-B-Translation_Myc" = "Group 3/4-B-Translation_Myc",
   "Northcutt_Group 3/4-C-Differentiation" = "Group 3/4-C-Differentiation",
  "CytoTRACE-GCS" = "GCS",
  "CytoTRACE-score" = "CytoTRACE",
  unaligned_cluster_diversity = "unaligned_cluster_diversity", 
  aligned_cluster_diversity = "aligned_cluster_diversity",
  `Subpopulations` = "tumor_subpopulations_gp4",
  `CNV count (methylation)` = "methyl_nCNV"
)

so@assays$alra <- get_alra_assay(so,
                                 file.path(alra_objs, "gp4.qs"),
                                 overwrite = FALSE)

dir.create(file.path(cb_outdir, "human", "GP4-neoplastic-cells"),
           showWarnings = FALSE,
           recursive = TRUE)
           
markdown::markdownToHTML(text = c(
                   "Neoplastic cells from Group 4 subgroup.  ",
                   "There are both unaligned (umap) and aligned (harmony_umap) projections  ",
                   "",
                   "##### Annotation description:  ",
                   "",
                   "* clusters unaligned: clusters derived from PCA + Louvain clustering",
                   "* clusters aligned: clusters derived from Harmony + Louvain clustering",
                   "* subtype: The highest correlated MB subtype based on classifications from Cavalli FMG, Remke M, Rampasek L, et al. Intertumoral Heterogeneity within Medulloblastoma Subgroups. Cancer Cell. 2017;31:737–754.e6",
                   "* UPN: Unique patient identifier",
                   "* cell cycle phase: inferred cell cycle phase",
                   "* Northcutt SHH-*: Gene modules generated from subpopulations described in Hovestadt V, Smith KS, Bihannic L, et al. Resolving medulloblastoma cellular architecture by single-cell genomics. Nature. 2019;572:74–79.",
                   "* Cytotrace-*: Scores derived by applying the cytotrace algorithm",
                   "* cluster diversity: Entropy measure of tumor diversity within each cluster, 0 = cluster derived from 1 patient, 1 = cluster equal mix from all patients.",
                   "* Subpopulations: subpopulations described in manuscript, these annotations are displayed in the browser, and the markers displayed match the harmony_umap layout",
                   "* CNV count (methylation): # of CNVs based on methylation data  ",
                   "",
                   "##### Feature descriptions",
                   "",
                   "* CNV calls per cell based on inferCNV have been added to the data matrix, which can be viewed by querying the 'Color by gene' tab for cnv metrics. The metrics are prefixed with 'CNV_' (e.g. CNV_has-cnv-chr9, CNV_proportion_cnv-chr22)  ",
                   "  +  The fields has_cnv/has_loss/has_dupli are 0/1 boolean answers to whether respectively any kind of cnv, a loss or a duplication is found within the given chr.  ",
"  +  proportion_cnv/proportion_loss/proportion_dupli is the proportion in number of genes that are part of any cnv/loss cnv/duplication cnv in the given chr.  ",
"  +  proportion_scaled are similar to proportion but taking into account whether genes that are duplications/losses are of a single copy or two copies as a weight  ",
"  +  top_loss_n/top_dupli_n are the n loss/duplication cnvs that cover the most genes  ",
                   "* ALRA imputed values are also present for visualization purposes and can be queried using 'ALRA_gene_name' (e.g. ALRA_GRIA2) "),
  output = file.path(cb_outdir, "human", "GP4-neoplastic-cells", "abstract.html"),
  fragment.only = TRUE)


make_cellbrowser(so, 
                 column_list = cols_to_keep,
                 secondary_cols = 
                   c("subtype",
                     "cell_type",
                     "subgroup",
                     "cell cycle phase"), 
                 project = "GP4-neoplastic-cells",
                 outdir = file.path(cb_outdir, "human"),
                 marker_file = "markers/gp4/harmony_subpopulation_markers_gp4.tsv",
                 ident = "Subpopulations",
                 embeddings = def_embeddings(so),
                 skip_expr_matrix = TRUE,
                 assays = c("ALRA_" = "alra",
                            "CNV_" = "CNV"),
                config = list(priority = 6),
                 description = list(    
                   title = "Medulloblastoma Group 4 subgroup",
                   abstractFile = "abstract.html" 
                 )
)
```



## Immune populations


```{r}
so <- qread(file.path("objects", "immune", "immune_so.qs"))
so$subgroup <- ifelse(so$subgroup == "GP3/4", "GP3", so$subgroup)
cols_to_keep <- c(
  `genes per cell` = "nFeature_RNA",
  `UMIs per cell` = "nCount_RNA",
  `percent of UMIs as mitochondria` = "percent.mt",
  `clusters` = "fine_immune_clusters",
  `subgroup` = "subgroup",
  `other_sample_info` = "additional_info",
  `UPN` = "UPN",
  cell_types = "publication_cell_types")

so@assays$alra <- get_alra_assay(so,
                                 file.path(alra_objs, "immune.qs"),
                                 overwrite = FALSE)

dir.create(file.path(cb_outdir, "human", "Immune-cells"),
           recursive = T,
           showWarnings = F)
markdown::markdownToHTML(text = c(
                   "Myleoid and lymphocyte populations from all subgroups.",
                   "There is only an unaligned projection, as alignment was not necessary ",
                   "",
                   "##### Annotation descriptions:  ",
                   "",
                   "* subgroup: Tumor subgroup",
                   "* UPN: Unique patient identifier",
                   "* Other sample info: Information on primary or recurrence status",
                   "* cell types: Inferred immune cell type"
                   ),
  output = file.path(cb_outdir, "human", "Immune-cells", "abstract.html"),
  fragment.only = TRUE)

make_cellbrowser(so, 
                 column_list = cols_to_keep,
                 secondary_cols = c(
                   "subgroup"), 
                 project = "Immune-cells",
                 outdir = file.path(cb_outdir, "human"),
                 marker_file = "markers/immune/immune_publication_cell_type_markers.tsv",
                 ident = "cell_types",
                 embeddings = def_embeddings(so),
                 skip_expr_matrix = TRUE,
                 assays = c("ALRA_" = "alra"),
                 config = list(priority = 7),
                 description = list(    
                   title = "Medulloblastoma immune populations",
                   abstractFile = "abstract.html")
)
```

# Mouse models

```{r}
collection_cb <- c(shortLabel = "Mouse MB tumor models (SHH and GP3)")

collection_desc <- c(title = "Medulloblastoma mouse models",
                     abstract = "Mouse tumors derived from SHH or GP3 models")

write_config(collection_cb, file.path(cb_outdir, "mouse", "cellbrowser.conf"))
write_config(collection_desc, file.path(cb_outdir, "mouse", "desc.conf"))
```

## Mouse model (Gp3) Myc-dnp53

```{r}
so <- qread(file.path("objects",
                  "mouse",
                  "gp34",
                  "gp3_myc_dnp53_sobj.qs"))

cols_to_keep <- c(
  `genes per cell` = "nFeature_RNA",
  `UMIs per cell` = "nCount_RNA",
  `percent of UMIs as mitochondria` = "percent.mt",
  `clusters` = "clusters",
  `sample` = "expt_ids",
   cell_type = "publication_cell_types",
   `cell cycle phase` = "Phase",
   `egfp expression` = "egfp",
  "Northcutt_cell_cycle_A" = "Group 3/4-A_cell_cycle",
  "Northcutt_Translation_Myc_B" = "Group 3/4-B_Translation_Myc",
  "Northcutt_Differentiation_C" = "Group 3/4-C_Differentiation"
  )

dir.create(file.path(cb_outdir, "mouse", "GP3-Myc-dnP53"),
           recursive = T,
           showWarnings = F)
markdown::markdownToHTML(text = c(
                   "scRNA-seq on GP3 mouse model (Myc + dnp53)",
                   "There is only an unaligned projection",
                   "",
                   "##### Annotation descriptions:  ",
                   "",
                   "* cell cycle phase: Inferred cell cycle phase",
                   "* cell type: Inferred cell subpopulation",
                   "* egfp expression: trangenic eGFP expression",
                   "* Northcutt SHH-*: Gene modules generated from subpopulations described in Hovestadt V, Smith KS, Bihannic L, et al. Resolving medulloblastoma cellular architecture by single-cell genomics. Nature. 2019;572:74–79."),
  output = file.path(cb_outdir, "mouse", "GP3-Myc-dnP53", "abstract.html"),
  fragment.only = TRUE)

make_cellbrowser(so, 
                 column_list = cols_to_keep,
                 secondary_cols = c(
                    "sample",
                    "cell cycle phase"), 
                 project = "GP3-Myc-dnP53",
                 outdir = file.path(cb_outdir, "mouse"),
                 marker_file = "markers/mouse/gp34/myc_dnp53_pub_cell_types_cluster_markers.tsv",
                 ident = "cell_type",
                 embeddings = def_embeddings(so),
                 skip_expr_matrix = TRUE,
                 description = list(    
                   title = "MYC + DNp53 Group 3 medulloblastoma mouse model ",
                   abstractFile = "abstract.html"),
                 config = list(priority = 1))
```

## Mouse model (Gp3) Myc-Gfi1
```{r}
so <- qread(file.path("objects",
                  "mouse",
                  "gp34",
                  "gp3_myc_gfi1_sobj.qs"))

cols_to_keep <- c(
  `genes per cell` = "nFeature_RNA",
  `UMIs per cell` = "nCount_RNA",
  `percent of UMIs as mitochondria` = "percent.mt",
  `clusters` = "clusters",
  `sample` = "expt_ids",
   cell_type = "publication_cell_types",
   `cell cycle phase` = "Phase",
   firefly_expression = "luc-firefly",
  "Northcutt_cell_cycle_A" = "Group 3/4-A_cell_cycle",
  "Northcutt_Translation_Myc_B" = "Group 3/4-B_Translation_Myc",
  "Northcutt_Differentiation_C" = "Group 3/4-C_Differentiation"
  )


dir.create(file.path(cb_outdir, "mouse", "GP3-Myc-Gfi1"),
           recursive = T,
           showWarnings = F)
markdown::markdownToHTML(text = c(
                   "scRNA-seq on GP3 mouse model (Myc + Gfi1)",
                   "There is only an unaligned projection",
                   "",
                   "##### Annotation descriptions:  ",
                   "",
                   "* cell cycle phase: Inferred cell cycle phase",
                   "* cell type: Inferred cell subpopulation",
                   "* firefly expression: trangenic luciferase expression",
                   "* Northcutt SHH-*: Gene modules generated from subpopulations described in Hovestadt V, Smith KS, Bihannic L, et al. Resolving medulloblastoma cellular architecture by single-cell genomics. Nature. 2019;572:74–79."),
  output = file.path(cb_outdir, "mouse", "GP3-Myc-Gfi1", "abstract.html"),
  fragment.only = TRUE)

make_cellbrowser(so, 
                 column_list = cols_to_keep,
                 secondary_cols = c(
                    "sample",
                    "cell cycle phase"), 
                 project = "GP3-Myc-Gfi1",
                 outdir = file.path(cb_outdir, "mouse"),
                 marker_file = "markers/mouse/gp34/myc_gfi1_pub_cell_types_cluster_markers.tsv",
                 ident = "cell_type",
                 embeddings = def_embeddings(so),
                 skip_expr_matrix = TRUE,
                 description = list(    
                   title = "MYC + GFI1 Group 3 medulloblastoma mouse model ",
                   abstractFile = "abstract.html"),
                 config = list(priority = 4))
```


## Mouse model (SHH) Math1-Cre1 SmoM2

```{r}
so <- qread(file.path("objects",
                  "mouse",
                  "shh",
                  "shh_sobj.qs"))


cols_to_keep <- c(
  `genes per cell` = "nFeature_RNA",
  `UMIs per cell` = "nCount_RNA",
  `percent of UMIs as mitochondria` = "percent.mt",
  `clusters` = "clusters",
  `sample` = "UPN",
   cell_type = "publication_cell_types",
   `cell cycle phase` = "Phase",
  "Northcutt_cell_cycle_A" = "SHH-A-cell_cycle",
  "Northcutt_SHH_signaling_B" = "SHH-B-SHH_signaling",
  "Northcutt_Differentiation_C" = "SHH-C-Differentiation"
  )


dir.create(file.path(cb_outdir, "mouse", "SHH-Math-Cre-SmoM2"),
           recursive = T,
           showWarnings = F)
markdown::markdownToHTML(text = c(
                   "scRNA-seq of SHH mouse model",
                   "There is only an unaligned projection",
                   "",
                   "##### Annotation descriptions:  ",
                   "",
                   "* sample: replicate samples",
                   "* cell cycle phase: Inferred cell cycle phase",
                   "* cell type: Inferred cell subpopulation",
                   "* Northcutt SHH-*: Gene modules generated from subpopulations described in Hovestadt V, Smith KS, Bihannic L, et al. Resolving medulloblastoma cellular architecture by single-cell genomics. Nature. 2019;572:74–79."),
  output = file.path(cb_outdir, "mouse", "SHH-Math-Cre-SmoM2", "abstract.html"),
  fragment.only = TRUE)


make_cellbrowser(so, 
                 column_list = cols_to_keep,
                 secondary_cols = c(
                    "sample",
                    "cell cycle phase"), 
                 project = "SHH-Math-Cre-SmoM2",
                 outdir = file.path(cb_outdir, "mouse"),
                 marker_file = "markers/mouse/shh/shh_pub_cell_types_cluster_markers.tsv",
                 ident = "cell_type",
                 embeddings = def_embeddings(so),
                 skip_expr_matrix = TRUE,
                 description = list(    
                   title = "SHH medulloblastoma mouse model ",
                   abstractFile = "abstract.html"),
                 config = list(priority = 7))
```



# Build all 

```{r, eval = FALSE}
cb_conf <- read_lines("~/.cellbrowser.conf") %>% 
  str_subset("dataRoot") %>% 
  str_split("=") %>% 
  .[[1]] %>% 
  .[2] %>% 
  gsub('\"', "", .)

if((Sys.getenv("CBDATAROOT") != cb_outdir) && (cb_outdir != cb_conf )) {
  write_lines(paste0('dataRoot="', 
                     normalizePath(cb_outdir), '"'),
              "~/.cellbrowser.conf")
}
```

```{r, eval = FALSE}
datasets <- c(
   "human/All-cells",
   "human/Immune-cells",
   "human/GP3-neoplastic-cells",
   "human/GP4-neoplastic-cells",
   "human/SHH-neoplastic-cells",
   "mouse/GP3-Myc-dnP53",
   "mouse/GP3-Myc-Gfi1",
   "mouse/SHH-Math-Cre-SmoM2"
)

datasets <- file.path(cb_outdir, datasets, "cellbrowser.conf")

build_cellbrowser(datasets, 
                    outdir = file.path(cb_outdir, "mblast"),
                    cbBuild_path =  "/miniconda3/bin/cbBuild")
```





